---
title: "clustering"
output: html_document
---

this (in theory) is gonna make a clustered heatmap of PSI for all the dAS genes. Briefly, dAS genes were identified as detected by both majiq and vast in at least one comparison (sashimi scoring indicated that events predicted by vast and majiq were well reflected in genomic alignments). For all dAS genes, we identified the PSI predicted by vast (but again, sashimi scoring showed that vast and majiq line up well and the vast output is easier to work with) for each event of each dAS gene in each treatment group. We then performed k-means clustering, plotted a heatmap of the PSI for each event in each cluster and performed GO enrichment analysis on each cluster.


later on we also tried grouping the treatments together, using both PCA and hierarchical clustering to see what patterns there are (Anna does a similar analysis in a committee report, and Calarco suggested it to me as well)
```{r}
library(ComplexHeatmap)
library(ggVennDiagram)
library(ggtext)
library(MASS)
library(spatstat)
library(ggrepel)
library(psych)
library(patchwork)
library(RColorBrewer)
library(palettetown)
library(tidyverse)
```

vast-tools:
```{r}
vast_k20dk20w <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/K20d-K20w_68events/K20d-K20w_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("K20dK20w"))

vast_k20dl20d <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/K20d-L20d_327events/K20d-L20d_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("K20dL20d"))

vast_k20wl20w <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/K20w-L20w_196events/K20w-L20w_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("K20wL20w"))

vast_k40dk40w <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/K40d-K40w_23events/K40d-K40w_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("K40dK40w"))

vast_k40dl40d <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/K40d-L40d_168events/K40d-L40d_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("K40dL40d"))

vast_k40wl40w <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/K40w-L40w_69events/K40w-L40w_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("K40wL40w"))

vast_k60dk60w <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/K60d-K60w_6events/K60d-K60w_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("K60dK60w"))

vast_k60dl60d <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/K60d-L60d_94events/K60d-L60d_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("K60dL60d"))

vast_k60wl60w <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/K60w-L60w_58events/K60w-L60w_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("K60wL60w"))

vast_l20dl20w <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/L20d-L20w_706events/L20d-L20w_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("L20dL20w"))

vast_l40dl40w <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/L40d-L40w_27events/L40d-L40w_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("L40dL40w"))

vast_l60dl60w <- read_delim("../../../vasttools/all_data/dPSI25_no_merging_all_reps/L60d-L60w_5events/L60d-L60w_compare_ouput_complete.tab", col_names = T, delim = "\t") %>% 
  select(GeneID, EVENT.x) %>% 
  mutate(prog = c("vast-tools"), comp = c("L60dL60w"))
```

majiq:
```{r}
majiq_k20dk20w <- read_delim("../../../majiq/dAS_analysis/filter25/K20dK20w.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("K20dK20w"))

majiq_k40dk40w <- read_delim("../../../majiq/dAS_analysis/filter25/K40dK40w.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("K40dK40w"))

majiq_k60dk60w <- read_delim("../../../majiq/dAS_analysis/filter25/K60dK60w.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("K60dK60w"))

majiq_l20dl20w <- read_delim("../../../majiq/dAS_analysis/filter25/L20dL20w.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("L20dL20w"))

majiq_l40dl40w <- read_delim("../../../majiq/dAS_analysis/filter25/L40dL40w.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("L40dL40w"))

majiq_l60dl60w <- read_delim("../../../majiq/dAS_analysis/filter25/L60dL60w.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("L60dL60w"))

majiq_k20dl20d <- read_delim("../../../majiq/dAS_analysis/filter25/K20dL20d.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("K20dL20d"))

majiq_k40dl40d <- read_delim("../../../majiq/dAS_analysis/filter25/K40dL40d.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("K40dL40d"))

majiq_k60dl60d <- read_delim("../../../majiq/dAS_analysis/filter25/K60dL60d.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("K60dL60d"))

majiq_k20wl20w <- read_delim("../../../majiq/dAS_analysis/filter25/K20wL20w.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("K20wL20w"))

majiq_k40wl40w <- read_delim("../../../majiq/dAS_analysis/filter25/K40wL40w.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("K40wL40w"))

majiq_k60wl60w <- read_delim("../../../majiq/dAS_analysis/filter25/K60wL60w.tsv", delim = "\t") %>% 
  select(`Gene ID`) %>% 
  mutate(prog = c("majiq"), comp = c("K60wL60w"))
```

pull out the vast events where dAS of that gene was also identified by maijq
```{r}
vm_k20dk20w <- filter(vast_k20dk20w, GeneID %in% majiq_k20dk20w$`Gene ID`)
vm_k40dk40w <- filter(vast_k40dk40w, GeneID %in% majiq_k40dk40w$`Gene ID`)
vm_k60dk60w <- filter(vast_k60dk60w, GeneID %in% majiq_k60dk60w$`Gene ID`)
vm_l20dl20w <- filter(vast_l20dl20w, GeneID %in% majiq_l20dl20w$`Gene ID`)
vm_l40dl40w <- filter(vast_l40dl40w, GeneID %in% majiq_l40dl40w$`Gene ID`)
vm_l60dl60w <- filter(vast_l60dl60w, GeneID %in% majiq_l60dl60w$`Gene ID`)
vm_k20dl20d <- filter(vast_k20dl20d, GeneID %in% majiq_k20dl20d$`Gene ID`)
vm_k40dl40d <- filter(vast_k40dl40d, GeneID %in% majiq_k40dl40d$`Gene ID`)
vm_k60dl60d <- filter(vast_k60dl60d, GeneID %in% majiq_k60dl60d$`Gene ID`)
vm_k20wl20w <- filter(vast_k20wl20w, GeneID %in% majiq_k20wl20w$`Gene ID`)
vm_k40wl40w <- filter(vast_k40wl40w, GeneID %in% majiq_k40wl40w$`Gene ID`)
vm_k60wl60w <- filter(vast_k60wl60w, GeneID %in% majiq_k60wl60w$`Gene ID`)
vm_allcomps <- rbind(vm_k20dk20w, vm_k20dl20d, vm_k20wl20w, vm_k40dk40w, vm_k40dl40d, vm_k40wl40w, vm_k60dk60w, vm_k60dl60d, vm_k60wl60w, vm_l20dl20w, vm_l40dl40w, vm_l60dl60w)

#this is useful for candidate genes, let's write it out
#write.table(vm_allcomps, "vm_all_dAS_genes.tsv", col.names = T, row.names = F, quote = F, sep = "\t")
```

let's plot this real quick:
```{r}
order1 <- c("K60dK60w", "K40dK40w", "K20dK20w", "L60dL60w", "L40dL40w", "L20dL20w")
order2 <- c("K60dL60d", "K40dL40d", "K20dL20d", "K60wL60w", "K40wL40w", "K20wL20w")

barplot1 <- vm_allcomps %>% filter(comp %in% order1) %>% 
  group_by(comp) %>% 
  summarize(count = n()) 
#add L60dL60w back in (just to make an empty bar)
L60dL60w <- tibble(comp = "L60dL60w", count = 0)
barplot1 <- rbind(barplot1, L60dL60w)

barplot1 <- barplot1 %>% mutate(comp = factor(comp, levels = order1)) %>% 
  arrange(comp)

barplot2 <- vm_allcomps %>% filter(comp %in% order2) %>% 
  group_by(comp) %>% 
  summarize(count = n()) %>%
  mutate(comp = factor(comp, levels = order2)) %>% 
  arrange(comp)

```

```{r}
png("test.png", units="in", width=6, height=4, res=300)
ggplot(barplot1, aes(x = comp, y = count, fill = count))+
  geom_bar(stat = "identity", fill = "darkturquoise")+
  labs(y = "Number of dAS genes\n")+
  theme_minimal()+
  theme(axis.title.x = element_blank())

dev.off()
```

```{r}
png("test.png", units="in", width=6, height=4, res=300)
ggplot(barplot2, aes(x = comp, y = count, fill = count))+
  geom_bar(stat = "identity", fill = "darkturquoise")+
  labs(y = "Number of dAS genes\n")+
  theme_minimal()+
  theme(axis.title.x = element_blank())

dev.off()
```

at this point, we have a list of all the dAS events we're interested in. many of them are obviously gonna be duplicates between comps (dAS in multiple comparisons), and since we're looking across all treatment groups anyway, we should just get rid of these now. All we're really interested in is the event IDs cause that's what we'll be looking for in the inclusion table. Should note here that some of these events may not have sufficient read coverage across all treatment groups, but they at least did in the 2 groups where the dAS was identified. 

```{r}
events <- unique(vm_allcomps$EVENT.x)
#ok, so there are 613 unique splicing events identified by vast and majiq. why is this less than the 643 we see in the Venn diagram? Because the same gene in different groups would've been counted as different observations there, but here they'd be the same 
```

The next step is to get out the vast inclusion table, and pull out the events corresponding to these genes. Isn't there a version of the inclusion table with AGI codes and replicate averages somewhere?
...apparently not
```{r}
inc <- read_delim("C:/Users/hasna/Documents/MSc_project/vasttools/all_data/complete_inclusion_table.tab", delim = "\t") 
db <- read_delim("C:/Users/hasna/Documents/MSc_project/vasttools/vast-tools/Ath/FILES/Ath.Event-Gene.IDs.txt", delim = "\t")
colnames(db) <- c("EVENT", "GeneID") #gotta get the colnames to match across DFs
inc <- merge(inc, db, by = "EVENT", all.x = T)

#filter the inclusion table to just the events we care about. this will make it much smaller and quicker to work with

inc <- filter(inc, EVENT %in% events)
```

next thing is to average the PSI across replicates. I'm thinking to get rid of the quality scores column, and then pivot longer, and then add a column for treatment group, and then group by treatment and event and summarize the mean?

```{r}
inc <- inc %>% select(!ends_with("-Q"), -c("GENE", "COORD", "LENGTH", "FullCO", "COMPLEX")) %>% 
  pivot_longer(cols = -c(EVENT, GeneID), names_to = "sample", values_to = "PSI") %>% 
  mutate("group" = case_when(grepl("K20d", sample) ~ "K20d",
                             grepl("K40d", sample) ~ "K40d",
                             grepl("K60d", sample) ~ "K60d",
                             grepl("L20d", sample) ~ "L20d",
                             grepl("L40d", sample) ~ "L40d",
                             grepl("L60d", sample) ~ "L60d",
                             grepl("K20W", sample) ~ "K20w",
                             grepl("K40W", sample) ~ "K40w",
                             grepl("K60W", sample) ~ "K60w",
                             grepl("L20W", sample) ~ "L20w",
                             grepl("L40W", sample) ~ "L40w",
                             grepl("L60W", sample) ~ "L60w",
                             grepl("Lrw", sample) ~ "Lrw",
                             grepl("Krw", sample) ~ "Krw")) %>% 
  group_by(EVENT, GeneID, group) %>% 
  summarize(mean = mean(PSI), sd = sd (PSI)) %>% 
  mutate("AS_percent" = ifelse(
  grepl("AthEX.*", EVENT), 100 - mean, mean
  ))
inc$mean <- NULL
inc_long <- inc
inc$sd <- NULL

#now we have to pivot wider again for k means
#inc <- inc %>% pivot_wider(names_from = group, values_from = AS_percent)

#pivot wider the other way for hclust
inc$GeneID <- NULL
inc <- inc %>% pivot_wider(names_from = EVENT, values_from = AS_percent)
```

that's all set up then, next thing is k-means clustering. let's start with a scree plot
```{r}
#first, let's write a function to test several different possible numbers of clusters in kmeans analysis
ss <- function(i, d) {
  d %>% ungroup %>% 
    select(where(is.numeric)) %>% 
    kmeans(i, nstart = 20) -> km
  km$tot.withinss
}
#next, construct a dataframe of cluster numbers (let's try 3-30 to start) and the corresponding tot.withinss (the total within cluster sum of squares, which is basically a measure of how tight the clusters are)
tibble(clusters = 2:20) %>% 
   mutate(tot.wss = map_dbl(clusters, ~ss(., inc))) -> scree

#plot scree_data. We're looking for the value of clusters where increasing cluster no longer sharply reduces total within cluster sum of squares
ggplot(scree, aes(x = clusters, y = tot.wss))+
  geom_point()+
  geom_line()
```

let's go with 6 clusters for now
```{r}
set.seed(43)
km6 <- inc %>% ungroup %>% 
  select(-c(GeneID, EVENT)) %>% 
  kmeans(6, nstart = 20)

inc.km <- tibble(EVENT = inc$EVENT, GeneID = inc$GeneID, cluster = km6$cluster)
#add the cluster numbers to the PSI table (inc)

inc6 <- merge(inc, inc.km, by = "EVENT", all.x = T)
```

make the heatmap: like Anna's, we want the columns split by treatment group, and the rows by cluster
```{r}
mat <- as.matrix(inc6[,3:16])
rownames(mat) <- inc6$cluster

Heatmap(mat,  row_split = factor(inc6$cluster, levels = 6:1), cluster_row_slices = F, name = "Percent AS", show_row_dend = F, show_column_dend = F, show_row_names = F, column_order = c("K60d", "K60w", "K40d", "K40w", "K20d", "K20w", "Krw", "L60d", "L60w", "L40d", "L40w", "L20d", "L20w", "Lrw"))
```

looking at the heatmap i'm wondering if 6 clusters was too many (1 and 2 basically look the same). let's try 5
```{r}
set.seed(43)
km5 <- inc %>% ungroup %>% 
  select(-c(GeneID, EVENT)) %>% 
  kmeans(5, nstart = 20)

inc.km5 <- tibble(EVENT = inc$EVENT, GeneID = inc$GeneID, cluster = km5$cluster)
#add the cluster numbers to the PSI table (inc)

inc5 <- merge(inc, inc.km5, by = "EVENT", all.x = T)

mat5 <- as.matrix(inc5[,3:16])
rownames(mat5) <- inc5$cluster

text_list = list(
    text1 = "5",
    text2 = "4",
    text3 = "3",
    text4 = "2",
    text5 = "1"
)

ha = rowAnnotation(foo = anno_empty(border = FALSE, 
    width = max_text_width(unlist(text_list)) + unit(4, "mm")))
```

make the heatmap: like Anna's, we want the columns split by treatment group, and the rows by cluster
```{r}
png("test.png", units="in", width=12, height=12, res=300)

Heatmap(mat5,  row_split = factor(inc5$cluster, levels = 5:1), cluster_row_slices = F, name = "Percent AS\n", show_row_dend = F, show_column_dend = F, show_row_names = F, column_order = c("K60d", "K60w", "K40d", "K40w", "K20d", "K20w", "Krw", "L60d", "L60w", "L40d", "L40w", "L20d", "L20w", "Lrw"), column_names_rot = 0, column_names_centered = T, left_annotation = ha, row_title = "Cluster", column_names_gp = grid::gpar(fontsize = 16), row_title_gp = gpar(fontsize = 16), heatmap_legend_param = list(labels_gp = gpar(fontsize = 16), title_gp = gpar(fontsize = 16), legend_height = unit(4, "cm")))

for(i in 1:5) {
    decorate_annotation("foo", slice = i, {
        grid.rect(x = 0, width = unit(2, "mm"), gp = gpar(fill = i, col = NA), just = "left")
        grid.text(paste(text_list[[i]], collapse = "\n"), x = unit(3, "mm"), just = "left", gp = gpar(fontsize = 16))
    })
}

dev.off()
```

Nick wants to know where the K20dK20w dAS genes fall on the heatmap. let's figure out what cluster they're in 
```{r}
K20dK20w <- unique(vm_k20dk20w$GeneID)
K20dK20w_clusters <- inc5 %>% filter(GeneID.x %in% K20dK20w) %>% 
  select(cluster) %>% 
  group_by(cluster) %>% 
  summarize(count = n())
K20dK20w_clusters
```
aight so they're mostly in 3 and 5. we could also try replicting the heatmap with only these genes to get that little snapshot, but tbh honest I'm not sure it's worth it. 

looks good! next thing is GO enrichment on the clusters
```{r}
clust1 <- filter(inc5, cluster == 1) %>% 
  select(GeneID.x)
clust2 <- filter(inc5, cluster == 2) %>% 
  select(GeneID.x)
clust3 <- filter(inc5, cluster == 3) %>% 
  select(GeneID.x)
clust4 <- filter(inc5, cluster == 4) %>% 
  select(GeneID.x)
clust5 <- filter(inc5, cluster == 5) %>% 
  select(GeneID.x)  
write_csv(clust1, "clust1.csv")
write_csv(clust2, "clust2.csv")
write_csv(clust3, "clust3.csv")
write_csv(clust4, "clust4.csv")
write_csv(clust5, "clust5.csv")
```

```{r}
go1 <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/clustering/clust1_out.txt", delim = "\t") %>% 
  select(term_type, Term, queryitem, FDR) %>% 
  mutate(cluster = 1) %>% 
  filter (FDR <= 0.05)
go3 <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/clustering/clust3_out.txt", delim = "\t") %>% 
  select(term_type, Term, queryitem, FDR) %>% 
  mutate(cluster = 3)%>% 
  filter (FDR <= 0.05)
go4 <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/clustering/clust4_out.txt", delim = "\t") %>% 
  select(term_type, Term, queryitem, FDR) %>% 
  mutate(cluster = 4)%>% 
  filter (FDR <= 0.05)
go5 <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/clustering/clust5_out.txt", delim = "\t") %>% 
  select(term_type, Term, queryitem, FDR) %>% 
  mutate(cluster = 5)%>% 
  filter (FDR <= 0.05)
go2 <- tibble(term_type = "C", Term = "chloroplast part", queryitem = NA, FDR = NA, cluster = 2)
#went with chloro so it passes later filtering, will be removed by ggplot anyway cause of NAs
go <- rbind(go1, go2, go3, go4, go5)
go$cluster <- as.factor(go$cluster)

colour_key <- tibble(term_type = c("F", "C", "P"), colour = c("blue", "green", "red"))

go <- merge(go, colour_key, by = "term_type")

#need this gross workaround for some reason
thingy <-  paste("<span style = 'color: ", go$colour, ";'>", go$Term, "<span>", sep = "")
go <- cbind(go, thingy= thingy)

#instead of the number of genes associated with each term for dot size, let's try using the percent of genes in each cluster (so that the smaller clusters don't just have a bunch of tiny dots)
go <- go %>% mutate(percent = case_when(
  cluster == 1 ~ queryitem / 106,
  cluster == 2 ~ queryitem / 55, 
  cluster == 3 ~ queryitem / 110, 
  cluster == 4 ~ queryitem / 148,
  cluster ==5 ~ queryitem / 194
))
```

```{r}
png("test.png", units="in", width=12, height=15, res=300)
ggplot(go, aes(x = fct_inorder(cluster), y = thingy, colour = log10(FDR), size = percent, label = cluster))+
  geom_point()+
  scale_colour_gradient(low = "blue", high = "red")+
  theme_bw()+
  theme(axis.text.y = element_markdown())+
  labs(x = "",
    y = "GO term")
dev.off()
```

let's try filtering for GO terms that only include keywords of interest:
```{r}
filtered <- filter(go, grepl("translation|ribo|response|photo|phospho|metabol|oxi|protein|thylakoid|resp|chloro|mito|structur", thingy))
```

```{r}
png("test.png", units="in", width=12, height=15, res=300)
ggplot(filtered, aes(x = fct_inorder(cluster), y = thingy, colour = log10(FDR), size = percent, label = cluster))+
  geom_point()+
  scale_colour_gradient(low = "blue", high = "red")+
  theme_bw()+
  theme(axis.text.y = element_markdown(size = 15),
        legend.text = element_text(size = 15),
        legend.title = element_text(size = 15))+
  labs(x = "",
    y = "GO term",
    size = "Percent of Genes in Cluster")
dev.off()
```

let's make a venn diagram of some of the more populated comps. specifically the developmental ones and the 20% SWC ones. since we can only have 7 sets, lets see in K40wL40w and K60wL60w are similar, cause they're basically the same size.
```{r}
K40wL40w <- unique(vm_k40wl40w$GeneID)
K60wL60w <- unique(vm_k60wl60w$GeneID)
w6040 <- list(K40dL40w, K60wL60w)
ggVennDiagram(w6040,
              label_color = "white",
              label_size = 3,
              label_alpha = 0,
              label = "percent",
              category.names = c("K40wL40w", "K60wL60w"),
              set_size = 4,
              edge_size = 3)
```
ok so apparently they're incredibly different
maybe then we'll just stick to all the 20%s

```{r}
K20dK20w <- unique(vm_k20dk20w$GeneID)
L20dL20w <- unique(vm_l20dl20w$GeneID)
K20dL20d <- unique(vm_k20dl20d$GeneID)
K20wL20w <- unique(vm_k20wl20w$GeneID)
twenty <- list(K20dK20w, L20dL20w, K20dL20d, K20wL20w)
ggVennDiagram(twenty,
              label_color = "white",
              label_size = 3,
              label_alpha = 0,
              label = "percent",
              category.names = c("K20dK20w", "L20dL20w", "K20dL20d", "K20wL20w"),
              set_size = 4,
              edge_size = 1)
```

ok so apparently these are all different too. arguably then, AS is highly specific and regulated (or our dataset is highly noisy...)
there's an interesting overlap between L20dL20w and K20dL20d, which is probably driven by L20d being a monster

we've identified 2 candidate genes for further study: XTH31 and SAD1. wanna make a line graph that shows relative isoform usage over progressive drought (percent IR isoform and rep isoform). First step is to ensure that each gene has only one splicing event that's significant (not a big deal if it's more, just need to be aware of it). then we can pull out the PSI for each GC drought timepoint. this'll be cool cause it'll really help show the progressive drought as a timecourse, which it is. for the sake of discussion, we should also figure out which clusters these genes are in (3 or 5 would be ideal, but we can explain it away regardless)
```{r}
xth31 <- filter(inc_long, GeneID == "AT3G44990")
sad1 <- filter(inc_long, GeneID == "AT5G48870")
#filter(inc5, GeneID.x == "AT3G44990") %>% select(cluster) #XTH31 in cluster 5
#filter(inc5, GeneID.x == "AT5G48870") %>%  select (cluster) #SAD1 in cluster 5
```
ok, so we know the XTH event is IR1 (vast and majiq have the same event). what's the SAD1 one? cause the figure should have the gene structures (a), the line graph (b) and the K20dK20w sashimi plot (c)
Looked up the event id, it's' IR3. so that's interesting, cause the potential to change the gene structure is less. in fact, i should do a protein domain figure like i did for XTH31 in the CSPB talk, and make that part of (a) (transcript structures and protein domains)

pull out the percent IR over time in GCs. we need to add sd to the inclusion table so we can put error bars on the graph: this is inc_long
```{r}
order <- c("Rep_abundance", "AS_percent")
xth31 <- xth31 %>% filter(group == "K60d" | group == "K40d" | group == "K20d") %>% 
  mutate("Rep_abundance" = 100 - AS_percent) %>% 
  mutate(timepoint = str_sub(group, 2, 3)) %>% 
  pivot_longer(c(AS_percent, Rep_abundance), names_to = "isoform", values_to = "abundance") %>%  mutate(isoform = factor(isoform, levels = order)) %>% 
  arrange(isoform)

```

```{r}
ggplot(xth31, aes(x = fct_rev(timepoint), y = abundance, colour = fct_inorder(isoform), group = fct_inorder(isoform)))+
  geom_point()+
  geom_line()+
  geom_errorbar(aes(ymin = abundance - sd, ymax = abundance + sd), width = 0.1)+
  labs(x = "Timepoint (% SWC)",
       y = "Abundance (%)")+
  scale_color_discrete(name = "Isoform", labels = c("XTH31", "XTH31 IR"))
```

```{r}
order <- c("Rep_abundance", "AS_percent")
sad1 <- sad1 %>% filter(group == "K60d" | group == "K40d" | group == "K20d") %>% 
  mutate("Rep_abundance" = 100 - AS_percent) %>% 
  mutate(timepoint = str_sub(group, 2, 3)) %>% 
  pivot_longer(c(AS_percent, Rep_abundance), names_to = "isoform", values_to = "abundance") %>%  mutate(isoform = factor(isoform, levels = order)) %>% 
  arrange(isoform)

```

```{r}
ggplot(sad1, aes(x = fct_rev(timepoint), y = abundance, colour = fct_inorder(isoform), group = fct_inorder(isoform)))+
  geom_point()+
  geom_line()+
  geom_errorbar(aes(ymin = abundance - sd, ymax = abundance + sd), width = 0.1)+
  labs(x = "Timepoint (% SWC)",
       y = "Abundance (%)")+
  scale_color_discrete(name = "Isoform", labels = c("SAD1", "SAD1 IR"))
```

PCA: run up until chunk 4, then stop (after that we filter for dAS genes only and average the reps, for PCA I want all of it)
then run this:
```{r}
inc_pca <- read_delim("C:/Users/hasna/Documents/MSc_project/vasttools/all_data/complete_inclusion_table.tab", delim = "\t") 
db <- read_delim("C:/Users/hasna/Documents/MSc_project/vasttools/vast-tools/Ath/FILES/Ath.Event-Gene.IDs.txt", delim = "\t")
colnames(db) <- c("EVENT", "GeneID") #gotta get the colnames to match across DFs
inc_pca <- merge(inc_pca, db, by = "EVENT", all.x = T)
```


```{r}
inc_pca <- inc_pca %>% select(!ends_with("-Q"), -c("GENE", "COORD", "LENGTH", "FullCO", "COMPLEX")) %>% 
  pivot_longer(cols = -c(EVENT, GeneID), names_to = "sample", values_to = "PSI") %>% 
  mutate("group" = case_when(grepl("K20d", sample) ~ "K20d",
                             grepl("K40d", sample) ~ "K40d",
                             grepl("K60d", sample) ~ "K60d",
                             grepl("L20d", sample) ~ "L20d",
                             grepl("L40d", sample) ~ "L40d",
                             grepl("L60d", sample) ~ "L60d",
                             grepl("K20W", sample) ~ "K20w",
                             grepl("K40W", sample) ~ "K40w",
                             grepl("K60W", sample) ~ "K60w",
                             grepl("L20W", sample) ~ "L20w",
                             grepl("L40W", sample) ~ "L40w",
                             grepl("L60W", sample) ~ "L60w",
                             grepl("Lrw", sample) ~ "Lrw",
                             grepl("Krw", sample) ~ "Krw"))%>% 
  mutate("AS_percent" = ifelse(
  grepl("AthEX.*", EVENT), 100 - PSI, PSI
  ))
inc_pca$GeneID <- NULL
inc_pca$PSI <- NULL
#now we have to pivot wider again for pca
inc_pca <- inc_pca %>% pivot_wider(names_from = EVENT, values_from = AS_percent)
pca <- inc_pca %>% select(-c(sample, group))
```

```{r}
#PCA1: remove events where all are NA
#can't seem to make this work for some reason, but it wasn't telling us anything anyway
find_all_na <- function(x) any(!is.na(x))
pca1 <- pca %>% select(where(find_all_na))
pca1 <- pca1[is.na(pca1)] <- 0
as_pca1 <- prcomp(pca1)
summary(as_pca1)
as_pca1$x
pca_data1 <- cbind(as_pca1$x, inc_pca$sample, inc_pca$group)
pca_data1 <- as.data.frame(pca_data1)
ggplot(pca_data1, aes(x = PC1, y = PC2, colour = V44))+
  geom_point(size = 6)
#aight, so it looks like there's too much variability for this to really be meaningful... PC1 is ~6% and PC2 is ~4%, so this graph captures very little actual information
```

```{r}
#PCA2: remove events where any rows are NA
find_any_na <- function(x) all(!is.na(x))
pca2 <- pca %>% select(where(find_any_na))
as_pca2 <- prcomp(pca2)
summary(as_pca2)
as_pca2$x
pca_data2 <- cbind(as_pca2$x, inc_pca$sample, inc_pca$group)
pca_data2 <- as.data.frame(pca_data2)
pca_data2$cell_type <- str_sub(pca_data2$V44, 1, 1)
pca_data2$condition <- as.factor(str_sub(pca_data2$V44, 2, 4))
ggplot(pca_data2, aes(x = PC1, y = PC2, colour = cell_type, shape = condition))+
  scale_shape_manual(values=1:nlevels(pca_data3$condition))+
  geom_point(size = 6)+
    theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())+
  labs(x = "PC1 (8%)",
       y = "PC2 (6%)")
```

```{r}
#PCA3: only dAS events (same ones we used for k-means)
events <- unique(vm_allcomps$EVENT.x)
pca3 <- subset(pca, select = events)
as_pca3 <- prcomp(pca3)
summary(as_pca3)
as_pca3$x
pca_data3 <- cbind(as_pca3$x, inc_pca$sample, inc_pca$group)
pca_data3 <- as.data.frame(pca_data3)
pca_data3$cell_type <- str_sub(pca_data3$V44, 1, 1)
pca_data3$condition <- as.factor(str_sub(pca_data3$V44, 2, 4))
ggplot(pca_data3, aes(x = PC1, y = PC2, colour = cell_type, shape = condition))+
  scale_shape_manual(values=1:nlevels(pca_data3$condition))+
  geom_point(size = 6)+
    theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())+
  labs(x = "PC1 (44%)",
       y = "PC2 (15%)")
```

ok, how about hierarchical clustering? this will be somewhat similar. we still need an input df where rows are samples and columns are genes, which is what we have in the pca dfs (pca1, 2 and 3). we then need to convert these to dissimilarity matrices and run hclust. I wonder if it makes more sense to average the replicates (as we did for k-means) or keep them separate. for now I'll try keeping them separate, but if it's too messy we might need to rethink this
```{r}
#let's just use the dAS events for now
hc3 <- pca3
row.names(hc3) <- str_sub(inc_pca$sample, 1, 7)
hc3 <- dist(hc3, method = "euclidean")
as_hc3 <- hclust(hc3, method = "ward.D" )
plot(as_hc3, cex = 0.6, hang = -1)
rect.hclust(as_hc3, 4)
```

that looks good! to cluster by replicate averages, go back up and run chunk 9 and 10 (hclust part)
```{r}
inc_dist <- inc
names <- inc_dist$group
inc_dist$group <- NULL
row.names(inc_dist) <- names
inc_dist <- dist(inc_dist, method = "euclidean")
as_inc_dist <- hclust(inc_dist, method = "ward.D" ) 
plot(as_inc_dist, cex = 0.6, hang = -1)
rect.hclust(as_inc_dist, 3)
```

