---
title: "AS positions and sequences"
output: html_document
date: '2023-01-12'
---

```{r}
library(phylotools)
library(Biostrings)
library(tidyverse)
library(ggsignif)
library(ggridges)
library(ggVennDiagram)
library(VennDiagram)
```

need: vastDB, Tair10 GFF, representative gene models, list of AS events

list of AS events:
```{r}
events <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/clustering/vm_all_dAS_genes.tsv", delim = "\t") %>% 
  select(EVENT.x, comp) %>% 
  mutate(comp_type = case_when(
    comp == "K20dK20w" | comp == "K40dK40w"| comp == "K60dK60w" | comp == "L20dL20w" | comp == "L40dL40w" | comp == "L60dL60w" ~ "stress",
    comp == "K20dL20d" | comp == "K40dL40d" | comp == "K60dL60d" | comp == "K20wL20w" | comp == "K40wL40w" | comp == "K60wL60w" ~ "developmental"))
colnames(events) <- c("event", "comp", "comp_type")
```

vastDB
```{r}
inc <- read_delim("C:/Users/hasna/Documents/MSc_project/vasttools/all_data/complete_inclusion_table.tab", delim = "\t") 
db <- read_delim("C:/Users/hasna/Documents/MSc_project/vasttools/vast-tools/Ath/FILES/Ath.Event-Gene.IDs.txt", delim = "\t")
colnames(db) <- c("EVENT", "GeneID") #gotta get the colnames to match across DFs
inc <- merge(inc, db, by = "EVENT", all.x = T)

#filter the inclusion table to just the events we care about. this will make it much smaller and quicker to work with

inc <- filter(inc, EVENT %in% events$event) %>% 
  separate(COORD, into = c("chr", "coords"), sep = ":") %>% 
  separate(coords, into = c("start", "end"), sep = "-") %>% 
  select(EVENT, chr, start, end, GeneID)

rm(db)
```

representative gene models
```{r}
rep_mods_all <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/AS_positions_sequences/rep_gene_mods.tsv", delim = "\t", skip = 3, col_names = F) %>%  
  mutate(mod = X1) %>% 
  select(mod) %>% 
  mutate(GeneID = str_sub(mod, 1, 9))

rep_mods <- filter(rep_mods_all, GeneID %in% inc$GeneID)
#this is less than the number of events since some genes have multiple events

inc <- left_join(inc, rep_mods)
```


GFF:
```{r}
gff <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/AS_positions_sequences/Tair10.gff", delim = "\t", col_names = F)

#need to filter down to just the genes we're interested in, and to just their representative isoforms. For gene, mRNA and protein objects, this is ID, and for UTRs, exons and CDSs, it's parent. Should be able to grepl this part out, and assign it to a new column that we can filter from.

ID <- filter(gff, X3 %in% c("mRNA", "protein"))
parent <- filter(gff, X3 %in% c("exon", "CDS", "five_prime_UTR", "three_prime_UTR"))

ID$model <- str_sub(ID$X9, 4, 14)
parent$model <- str_sub(parent$X9, 8, 18)

gene_data <- rbind(ID, parent)
gene_data <- select(gene_data, X1, X3, X4, X5, X7, model) %>% 
  filter(model %in% rep_mods$mod)
colnames(gene_data) <- c("chr", "feat", "start", "end", "strand", "model")
```

```{r}
#for each event in inc, we want to get all the exonic positions, so pull all the exons for a gene and list out all numbers between the start and end positions, then order them (reverse them if they're on the minus strand) to get a list of all positions that make up the mRNA for that gene

pos_list <- data.frame()


for (i in 1:nrow(inc)){
  single_gene_inc <- inc[i, ]
  single_gene_data <- filter(gene_data, model == single_gene_inc$mod & feat == "exon") %>% 
    rowwise()
  positions = vector()
    for (j in 1:nrow(single_gene_data)){
      exon = single_gene_data[j, ]
      positions = c(positions, seq.int(exon$start, exon$end))
    }
  positions = data.frame(positions)
  positions <- positions %>% mutate(less = ifelse(positions <= single_gene_inc$start, 1, 0))
  pos_norm <- ifelse(single_gene_data[1, 5] == "-", 1 - (sum(positions$less))/(nrow(positions)), (sum(positions$less))/(nrow(positions)))
  pos_norm <- data.frame(gene = single_gene_inc$GeneID, event = single_gene_inc$EVENT, pos_norm = pos_norm)
  pos_list <- rbind(pos_list, pos_norm)
}

mrna_density <- density(pos_list$strand, adjust = 0.5, from = 0, to = 1)
mrna_density <- data.frame(x = mrna_density$x, y = mrna_density$y)
```

how well does this mirror splice junction locations (across the whole transcript length, are alt SJs distrbuted differently than all SJs?)
```{r}
exons <- filter(gff, X3 == "exon")
exons$model <- str_sub(exons$X9, 8, 18)

exons <- select(exons, X1, X3, X4, X5, X7, model) %>% 
  filter(model %in% rep_mods_all$mod)
colnames(exons) <- c("chr", "feat", "start", "end", "strand", "model")
exons2 <- exons %>% group_by(model) %>% 
  summarize(count = n()) %>% 
  filter(count >= 2) #22 146 genes have at least 2 exons (at least 1 SJ)

exons <- exons %>% filter(model %in% exons2$model)

#for each UTR, we can list out all the positions as before. The curveball is that now we have to identify the location of each splice junction in each UTR. The splice junction can be identified as the start of any exon other than the first one, so we need a system of numbering the exons as well. maybe we can filter out all the exons that aren't the first one and then run them individually? There's no real need to group junctions of the same gene together after all... 

exons <- exons %>% group_by(model) %>% 
  mutate(exon_num = row_number())

SJ <- filter(exons, exon_num != 1)
#so now the start positions in SJ define SJs, and we can run the same loop as for AS locations, where we piece together the exon positions, and identify the relative location of the junction

SJ_pos_list <- data.frame()


for (i in 1:nrow(SJ)){
  single_gene_inc <- SJ[i, ]
  single_gene_data <- filter(exons, model == single_gene_inc$model) %>% 
    rowwise()
  positions = vector()
    for (j in 1:nrow(single_gene_data)){
      temp = single_gene_data[j, ]
      positions = c(positions, seq.int(temp$start, temp$end))
    }
  positions = data.frame(positions)
  test_position <- ifelse(single_gene_data[1, 5] == "-", single_gene_inc$end, single_gene_inc$start)
  positions <- positions %>% mutate(less = ifelse(positions <= test_position[1,1], 1, 0))
  pos_norm <- ifelse(single_gene_data[1, 5] == "-", 1 - (sum(positions$less))/(nrow(positions)), (sum(positions$less))/(nrow(positions)))
  pos_norm <- data.frame(gene = single_gene_inc$model, pos_norm = pos_norm)
  SJ_pos_list <- rbind(SJ_pos_list, pos_norm)
}


junction_density <- density(SJ_pos_list$strand, adjust = 0.5, from = 0, to = 1)
junction_density <- data.frame(x = junction_density$x, y = junction_density$y)

#is this significantly different from AS location?
t.test(pos_list$strand, SJ_pos_list$strand, alternative = "less") #p = 0.048
ks.test(pos_list$strand, SJ_pos_list$strand, alternative = "less") #p = 0.0005
```

```{r}
png("test.png", width = 8, height = 3, units = "in", res = 300)
ggplot(junction_density, aes(x = x, fill = y, y = 1)) + 
  geom_tile(aes(height = 2)) +
  labs(x = "Normalized transcript location",
       y = "Density of splice junctions",
       legend = "Density")+
  theme(#axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  scale_fill_gradientn(colours = c("blue", "yellow", "red"), name = "Density")+
  geom_line(aes(y = y, x = x), colour = "black")
dev.off()
```


repeat the above for UTRs: some genes have introns in the UTR, so we'll keep the code the same (looping over each UTR exon) to handle this
```{r}
five_pos_list <- data.frame()

#some genes have no 5' UTR, need a way to skip them
five_utr <- filter(gene_data, feat == "five_prime_UTR")
inc_has_five <- filter(inc, mod %in% five_utr$model) #601 / 613

for (i in 1:nrow(inc_has_five)){
  single_gene_inc <- inc_has_five[i, ]
  single_gene_data <- filter(gene_data, model == single_gene_inc$mod & feat == "five_prime_UTR") %>% 
    rowwise()
  positions = vector()
    for (j in 1:nrow(single_gene_data)){
      five = single_gene_data[j, ]
      positions = c(positions, seq.int(five$start, five$end))
    }
  positions = data.frame(positions)
  positions <- positions %>% mutate(less = ifelse(positions <= single_gene_inc$start, 1, 0))
  pos_norm <- ifelse(single_gene_data[1, 5] == "-", 1 - (sum(positions$less))/(nrow(positions)), (sum(positions$less))/(nrow(positions)))
  pos_norm <- data.frame(gene = single_gene_inc$GeneID, event = single_gene_inc$EVENT, pos_norm = pos_norm)
  five_pos_list <- rbind(five_pos_list, pos_norm)
}

colnames(five_pos_list) <- c("gene", "event", "five_position")

#so anything with a 1 in this column has the AS after the 5' UTR, any non-one values are 5'UTR splicing

#repeat above for 3'UTR
three_pos_list <- data.frame()

#some genes have no 5' UTR, need a way to skip them
three_utr <- filter(gene_data, feat == "three_prime_UTR")
inc_has_three <- filter(inc, mod %in% three_utr$model) #610 / 613

for (i in 1:nrow(inc_has_three)){
  single_gene_inc <- inc_has_three[i, ]
  single_gene_data <- filter(gene_data, model == single_gene_inc$mod & feat == "three_prime_UTR") %>% 
    rowwise()
  positions = vector()
    for (j in 1:nrow(single_gene_data)){
      three = single_gene_data[j, ]
      positions = c(positions, seq.int(three$start, three$end))
    }
  positions = data.frame(positions)
  positions <- positions %>% mutate(less = ifelse(positions <= single_gene_inc$start, 1, 0))
  pos_norm <- ifelse(single_gene_data[1, 5] == "-", 1 - (sum(positions$less))/(nrow(positions)), (sum(positions$less))/(nrow(positions)))
  pos_norm <- data.frame(gene = single_gene_inc$GeneID, event = single_gene_inc$EVENT, pos_norm = pos_norm)
  three_pos_list <- rbind(three_pos_list, pos_norm)
}

colnames(three_pos_list) <- c("gene", "event", "three_position")
```

last thing is the CDS: get AS positions
```{r}
cds_pos_list <- data.frame()

for (i in 1:nrow(inc)){
  single_gene_inc <- inc[i, ]
  single_gene_data <- filter(gene_data, model == single_gene_inc$mod & feat == "CDS") %>% 
    rowwise()
  positions = vector()
    for (j in 1:nrow(single_gene_data)){
      cds = single_gene_data[j, ]
      positions = c(positions, seq.int(cds$start, cds$end))
    }
  positions = data.frame(positions)
  positions <- positions %>% mutate(less = ifelse(positions <= single_gene_inc$start, 1, 0))
  pos_norm <- ifelse(single_gene_data[1, 5] == "-", 1 - (sum(positions$less))/(nrow(positions)), (sum(positions$less))/(nrow(positions)))
  pos_norm <- data.frame(gene = single_gene_inc$GeneID, event = single_gene_inc$EVENT, pos_norm = pos_norm)
  cds_pos_list <- rbind(cds_pos_list, pos_norm)
}

colnames(cds_pos_list) <- c("gene", "event", "cds_position")
```

now to figure out if AS is in the UTRs or CDS
```{r}
utr_cds <- left_join(cds_pos_list, five_pos_list)
utr_cds <- left_join(utr_cds, three_pos_list)

utr_cds <- utr_cds %>% mutate(location = case_when(
  (cds_position == 1) ~ "three_prime_UTR",
  (cds_position != 1 & cds_position != 0) ~ "CDS",
  (cds_position == 0) ~ "five_prime_UTR"
))

nrow(filter(utr_cds, location == "CDS")) #585
nrow(filter(utr_cds, location == "five_prime_UTR")) #14
nrow(filter(utr_cds, location == "three_prime_UTR")) #14

utr_cds <- utr_cds %>% mutate(type = case_when(
  grepl("AthEX.*", event) ~ "Exon skipping",
  grepl("AthALTA.*", event) ~ "Alternative 3' SS",
  grepl("AthALTD", event) ~ "Alternative 5' SS",
  grepl("AthINT", event) ~ "Intron retention"
))

#ggplot(utr_cds, aes(x = location, fill = type))+
  #geom_bar(stat = "count")+
  #scale_fill_manual(values = c("grey70", "chartreuse2", "palevioletred1", "cadetblue2"))

utr_cds_comps <- left_join(utr_cds, events)
#filter(utr_cds, location == "three_prime_UTR") %>% 
 # group_by(comp_type) %>% 
  #summarize(count = n())

#what about the overlap between the 3 locations:
as_cds <- utr_cds %>% filter(location == "CDS") %>% 
  select(gene)
as_cds <- unique(as_cds$gene)
as_5utr <- utr_cds %>% filter(location == "five_prime_UTR") %>% 
  select(gene)
as_5utr <- unique(as_5utr$gene)
as_3utr <- utr_cds %>% filter(location == "three_prime_UTR") %>% 
  select(gene)
as_3utr <- unique(as_3utr$gene)
location <- list(as_3utr, as_cds, as_5utr)
```

venn diagram of AS locations (CDS, UTRs)
```{r}
png("test.png", width = 5, height = 5, units = "in", res = 300)
ggVennDiagram(location,
              label_color = "white",
              label_alpha = 0,
              category.names = c("3' UTR", "CDS", "5' UTR"),
              label_size = 4,
              set_size = 6,
              edge_size = 1)+
  scale_fill_gradient(high = "orchid", low = "grey")+
  theme(legend.position = "none")
dev.off()
```

```{r}
venn.diagram(x = location, filename = "test.png", category = c("3' UTR", "CDS", "5' UTR"), fontfamily = "sans", cat.fontfamily = "sans", ext.text = T, cat.pos = c(0, 0, 0), cat.cex = c(1.5, 1.5, 1.5), col = c("palegreen1", "limegreen", "lightgreen"), fill = c("palegreen1", "limegreen", "lightgreen"), overrideTriple = T)
```


now plot locations: cDNA density
```{r}
png("test.png", width = 8, height = 3, units = "in", res = 300)
ggplot(mrna_density, aes(x = x, fill = y, y = 1)) + 
  geom_tile(aes(height = 3)) +
  labs(x = "Normalized transcript location",
       y = "Density of AS start positions",
       legend = "Density")+
  theme(#axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  scale_fill_gradientn(colours = c("blue", "yellow", "red"), name = "Density")+
  geom_line(aes(y = y), colour = "black")
dev.off()
```

cDNA by comp_type (histograms)
```{r}
cdna_stress <- events %>% filter(comp_type == "stress") %>% 
  left_join(., pos_list, by = "event")

cdna_dev <- events %>% filter(comp_type == "developmental") %>% 
  left_join(., pos_list, by = "event")

#where, generally, are these things located?
nrow(filter(pos_list, strand < 0.5))/nrow(pos_list) #58.24% events before the halfway point
nrow(filter(cdna_stress, strand < 0.5))/nrow(cdna_stress) #61.62 % before 0.5
nrow(filter(cdna_dev, strand < 0.5))/nrow(cdna_dev) #53.35% before 0.5
t.test(cdna_stress$strand, cdna_dev$strand) #p = 0.0067 (**) by 2-sample t-test -> stress events happen earlier in the transcript

```

```{r}
png("test.png", units="in", width=6, height=4, res=300)
ggplot(cdna_stress, aes(x = strand))+
  geom_histogram(binwidth = 0.05, fill = "lightblue")+
  lims(x =c(0,1))+
  labs(x = "Normalized transcript location",
       y = "Stress-induced AS events \n(n = 383)")+
  theme_minimal()+
  theme(axis.ticks.y = element_blank())
dev.off()
```

```{r}
png("test.png", units="in", width=6, height=4, res=300)
ggplot(cdna_dev, aes(x = strand))+
  geom_histogram(binwidth = 0.05, fill = "lightsteelblue3")+
  lims(x =c(0, 1))+
  labs(x = "Normalized transcript location",
       y = "Developmental AS events \n(n = 328)")+
  scale_y_reverse()+
  theme_minimal()+
  theme(axis.ticks.y = element_blank())
dev.off()
```


plot CDS density
```{r}
cds <- filter(utr_cds, cds_position != 1 & cds_position != 0)
cds_density <- density(cds$cds_position, adjust = 0.5, from = 0, to = 1)
cds_density <- data.frame(x = cds_density$x, y = cds_density$y)
ggplot(cds_density, aes(x = x, fill = y, y = 1)) + 
  geom_tile(aes(height = 3)) +
  labs(x = "Normalized CDS location",
       y = "Density of AS start positions",
       legend = "Density")+
  theme(#axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  scale_fill_gradientn(colours = c("blue", "yellow", "red"), name = "Density")+
  geom_line(aes(y = y), colour = "black")
```

plot 5'UTR density
```{r}
five_prime <- filter(utr_cds, five_position != 1 & five_position != 0)
#five_density <- density(five_prime$five_position, adjust = 0.5, from = 0, to = 1)
five_density <- density(five_prime$five_position, adjust = 0.5)
five_density <- data.frame(x = five_density$x, y = five_density$y)
ggplot(five_density, aes(x = x, fill = y, y = 2)) + 
  geom_tile(aes(height = 5)) +
  labs(x = "Normalized 5' UTR location",
       y = "Density of AS start positions",
       legend = "Density")+
  theme(#axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  scale_fill_gradientn(colours = c("blue", "yellow", "red"), name = "Density")+
  geom_line(aes(y = y, x = x), colour = "black")
```

plot 3'UTR density
```{r}
three_prime <- filter(utr_cds, three_position != 1 & three_position != 0)
#three_density <- density(three_prime$three_position, adjust = 0.5, from = 0, to = 1)
three_density <- density(three_prime$three_position, adjust = 0.5)
three_density <- data.frame(x = three_density$x, y = three_density$y)
ggplot(three_density, aes(x = x, fill = y, y = 5)) + 
  geom_tile(aes(height = 10)) +
  labs(x = "Normalized 3' UTR location",
       y = "Density of AS start positions",
       legend = "Density")+
  theme(#axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  scale_fill_gradientn(colours = c("blue", "yellow", "red"), name = "Density")+
  geom_line(aes(y = y, x = x), colour = "black")
```

The cDNA and CDS look pretty uninteresting, but the UTRs could be kinda cool if the trend is real. However it might just be that all UTR SJs are located towards the CDS side. To figure that out, plot the density of UTR splice junctions

5' UTR
```{r}
fpu <- filter(gff, X3 == "five_prime_UTR")
fpu$model <- str_sub(fpu$X9, 8, 18)

fpu <- select(fpu, X1, X3, X4, X5, X7, model) %>% 
  filter(model %in% rep_mods_all$mod)
colnames(fpu) <- c("chr", "feat", "start", "end", "strand", "model")
fpu2 <- fpu %>% group_by(model) %>% 
  summarize(count = n()) %>% 
  filter(count >= 2)
max(fpu2$count) #ok wow so there's at least 1 gene with 9 exons in the 5' UTR... so the SJ code needs to account for more than 1 SJ. 3403 genes have at least 1 SJ in the 5' UTR

fpu <- fpu %>% filter(model %in% fpu2$model)

#for each UTR, we can list out all the positions as before. The curveball is that now we have to identify the location of each splice junction in each UTR. The splice junction can be identified as the start of any exon other than the first one, so we need a system of numbering the exons as well. maybe we can filter out all the exons that aren't the first one and then run them individually? There's no real need to group junctions of the same gene together after all... 

fpu <- fpu %>% group_by(model) %>% 
  mutate(exon_num = row_number())

junctions5 <- filter(fpu, exon_num != 1)
#so now the start positions in junctions5 define 5' UTR SJs, and we can run the same loop as for AS locations, where we piece together the UTR positions, and identify the relative location of the junction

junction5_pos_list <- data.frame()


for (i in 1:nrow(junctions5)){
  single_gene_inc <- junctions5[i, ]
  single_gene_data <- filter(fpu, model == single_gene_inc$model) %>% 
    rowwise()
  positions = vector()
    for (j in 1:nrow(single_gene_data)){
      five = single_gene_data[j, ]
      positions = c(positions, seq.int(five$start, five$end))
    }
  positions = data.frame(positions)
  test_position <- ifelse(single_gene_data[1, 5] == "-", single_gene_inc$end, single_gene_inc$start)
  positions <- positions %>% mutate(less = ifelse(positions <= test_position[1,1], 1, 0))
  pos_norm <- ifelse(single_gene_data[1, 5] == "-", 1 - (sum(positions$less))/(nrow(positions)), (sum(positions$less))/(nrow(positions)))
  pos_norm <- data.frame(gene = single_gene_inc$model, pos_norm = pos_norm)
  junction5_pos_list <- rbind(junction5_pos_list, pos_norm)
}


five_junction_density <- density(junction5_pos_list$strand, adjust = 0.5, from = 0, to = 1)
five_junction_density <- data.frame(x = five_junction_density$x, y = five_junction_density$y)
ggplot(five_junction_density, aes(x = x, fill = y, y = 1.5)) + 
  geom_tile(aes(height = 3)) +
  labs(x = "Normalized 5' UTR location",
       y = "Density of splice junctions",
       legend = "Density")+
  theme(#axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  scale_fill_gradientn(colours = c("blue", "yellow", "red"), name = "Density")+
  geom_line(aes(y = y, x = x), colour = "black")
```

```{r}
tpu <- filter(gff, X3 == "three_prime_UTR")
tpu$model <- str_sub(tpu$X9, 8, 18)

tpu <- select(tpu, X1, X3, X4, X5, X7, model) %>% 
  filter(model %in% rep_mods_all$mod)
colnames(tpu) <- c("chr", "feat", "start", "end", "strand", "model")
tpu2 <- tpu %>% group_by(model) %>% 
  summarize(count = n()) %>% 
  filter(count >= 2)
max(tpu2$count) #ok wow so there's at least 1 gene with 15 exons in the 3' UTR... so the SJ code needs to account for more than 1 SJ. 3403 genes have at least 1 SJ in the 5' UTR

tpu <- tpu %>% filter(model %in% tpu2$model)

tpu <- tpu %>% group_by(model) %>% 
  mutate(exon_num = row_number())

junctions3 <- filter(tpu, exon_num != 1)
#so now the start positions in junctions5 define 5' UTR SJs, and we can run the same loop as for AS locations, where we piece together the UTR positions, and identify the relative location of the junction

junction3_pos_list <- data.frame()


for (i in 1:nrow(junctions3)){
  single_gene_inc <- junctions3[i, ]
  single_gene_data <- filter(tpu, model == single_gene_inc$model) %>% 
    rowwise()
  positions = vector()
    for (j in 1:nrow(single_gene_data)){
      three = single_gene_data[j, ]
      positions = c(positions, seq.int(three$start, three$end))
    }
  positions = data.frame(positions)
  test_position <- ifelse(single_gene_data[1, 5] == "-", single_gene_inc$end, single_gene_inc$start)
  positions <- positions %>% mutate(less = ifelse(positions <= test_position[1,1], 1, 0))
  pos_norm <- ifelse(single_gene_data[1, 5] == "-", 1 - (sum(positions$less))/(nrow(positions)), (sum(positions$less))/(nrow(positions)))
  pos_norm <- data.frame(gene = single_gene_inc$model, pos_norm = pos_norm)
  junction3_pos_list <- rbind(junction3_pos_list, pos_norm)
}


three_junction_density <- density(junction3_pos_list$strand, adjust = 0.5, from = 0, to = 1)
three_junction_density <- data.frame(x = three_junction_density$x, y = three_junction_density$y)
ggplot(three_junction_density, aes(x = x, fill = y, y = 2.25)) + 
  geom_tile(aes(height = 4.5)) +
  labs(x = "Normalized 3' UTR location",
       y = "Density of splice junctions",
       legend = "Density")+
  theme(#axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  scale_fill_gradientn(colours = c("blue", "yellow", "red"), name = "Density")+
  geom_line(aes(y = y, x = x), colour = "black")
```
so AS density in the UTRs closely mirrors SJ density... no surprises there... 

for PTCs, we're only interested in IR events that happen in the CDS
```{r}
ir_cds <- filter(utr_cds, location == "CDS" & type == "Intron retention") #572 events
ir_cds_inc <- filter(inc, EVENT %in% ir_cds$event)
athal_gdna <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/AS_positions_sequences/Athaliana_167_TAIR10.fa.gz")
athal_gdna$seq.name <- str_sub(athal_gdna$seq.name, 1, 4)

ir_cds_seq <- data.frame()

for (i in 1:nrow(ir_cds_inc)){
  single_gene_inc <- ir_cds_inc[i, ]
  single_gene_data <- filter(gene_data, model == single_gene_inc$mod & feat == "CDS") 
  dir = single_gene_data[1, 5]
  single_gene_cds <-  select(single_gene_data, start, end)
  single_gene_pos <- select(single_gene_inc, start, end)
  complete_gene_data <- as.data.frame(rbind(single_gene_cds, single_gene_pos))
  sorted <- complete_gene_data %>% arrange(start)

  seq = data.frame()
  
  for (j in 1:nrow(sorted)){
      chr_num = single_gene_data[1,1]
      chr <- dplyr::filter(athal_gdna, seq.name == as.character(chr_num[1,1]))
      temp <- str_sub(chr$seq.text, sorted[j, 1], sorted[j, 2])
      temp <- paste(temp, collapse = "")
      seq[j, 1] <- temp
      temp <- vector()
  }
  seq <- as.vector(seq$V1)
  seq <- paste(seq, collapse = "")
  if (dir[1, 1] == "-"){
  seq <- as.character(reverseComplement(DNAString(seq)))
  }
  seq <- as.character(translate(DNAString(seq)))
  sequence <- data.frame(gene = single_gene_inc$GeneID, event = single_gene_inc$EVENT, sequence = seq)
  ir_cds_seq <- rbind(ir_cds_seq, sequence)
}

#write.table(ir_cds_seq, "C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/AS_positions_sequences/IR_CDS_sequences.tsv", row.names = F, quote = F)

ir_cds_seq <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/AS_positions_sequences/IR_CDS_sequences.tsv")
```

need to get lengths for the IR AA sequences (up to first *) and compare to representative isoforms to determine PTCs. then plot PTC-causing events.
```{r}
athal_cds <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Athaliana.cds.fa.gz") 
athal_cds$seq.name <- str_sub(athal_cds$seq.name, 1, 11)
athal_cds <- filter(athal_cds, seq.name %in% ir_cds_inc$mod)
athal_cds$rep_length <- nchar(athal_cds$seq.text)/3 - 1
athal_cds$model <- athal_cds$seq.name

ir_cds_seq$orf <- sub("\\*.*", "", ir_cds_seq$sequence)
ir_cds_seq$as_length <- nchar(ir_cds_seq$orf)
ir_cds_seq$model <- ir_cds_inc$mod
ir_cds_seq$start <- as.numeric(ir_cds_inc$start)
ir_cds_seq$end <- as.numeric(ir_cds_inc$end)
ir_cds_seq$intron_length <- abs(ir_cds_seq$end - ir_cds_seq$start)

ir_cds_seq <- left_join(ir_cds_seq, athal_cds[, c(3, 4)])
ir_cds_seq$ptc <- ifelse(ir_cds_seq$as_length < ir_cds_seq$rep_length + (ir_cds_seq$intron_length/3), "yes", "no")
nrow(filter(ir_cds_seq, ptc == "no")) #7 events have no PTC

ir_cds_seq <- left_join(ir_cds_seq, cds_pos_list[, c(2, 3)])

ir_cds_ptc <- filter(ir_cds_seq, ptc == "yes")
ir_cds_no_ptc <- filter(ir_cds_seq, ptc == "no")

ir_cds_ptc_density <- density(ir_cds_ptc$cds_position, adjust = 0.5, from = 0, to = 1)
ir_cds_ptc_density <- data.frame(x = ir_cds_ptc_density$x, y = ir_cds_ptc_density$y)
```

```{r}
ggplot(ir_cds_ptc_density, aes(x = x, fill = y, y = 1)) + 
  geom_tile(aes(height = 2)) +
  labs(x = "Normalized CDS location",
       y = "Density of PTC-causing IR start positions",
       legend = "Density")+
  theme(#axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  scale_fill_gradientn(colours = c("blue", "yellow", "red"), name = "Density")+
  geom_line(aes(y = y), colour = "black")
```

plot non-PTC events, and figure out how many of those are developmental
```{r}
ir_cds_no_ptc_density <- density(ir_cds_no_ptc$cds_position, adjust = 0.5, from = 0, to = 1)
ir_cds_no_ptc_density <- data.frame(x = ir_cds_no_ptc_density$x, y = ir_cds_no_ptc_density$y)
ggplot(ir_cds_no_ptc_density, aes(x = x, fill = y, y = 1.5)) + 
  geom_tile(aes(height = 3)) +
  labs(x = "Normalized CDS location",
       y = "Density of frame-preserving IR start positions",
       legend = "Density")+
  theme(#axis.text.y = element_blank(),
        axis.ticks.y = element_blank())+
  scale_fill_gradientn(colours = c("blue", "yellow", "red"), name = "Density")+
  geom_line(aes(y = y), colour = "black")

filter(events, event %in% ir_cds_no_ptc$event)
```

plot IR-PTC events by drought/dev comp (histograms)
```{r}
ptc_type <- left_join(ir_cds_seq, events)
ptc_stress <- filter(ptc_type, comp_type == "stress")
ptc_dev <- filter(ptc_type, comp_type == "developmental")
```

```{r}
png("test.png", units="in", width=8, height=4, res=300)
ggplot(ptc_stress, aes(x = cds_position))+
  geom_histogram(binwidth = 0.05, fill = "lightblue")+
  lims(x =c(0,1))+
  labs(x = "Normalized CDS location",
       y = "Stress-induced PTC-causing IR events \n(n = 357)")+
  theme(axis.ticks.y = element_blank())
dev.off()
```


```{r}
png("test.png", units="in", width=8, height=4, res=300)
ggplot(ptc_dev, aes(x = cds_position))+
  geom_histogram(binwidth = 0.05, fill = "lightsteelblue3")+
  lims(x =c(0, 1))+
  labs(x = "Normalized CDS location",
       y = "Developmental PTC-causing IR events \n(n = 306)")+
  scale_y_reverse()
dev.off()
```

for the IR events, get the cDNA sequence (or the CDS if AtRTD3 provides that for comparison)
scrap this, TSS are too variable for this to be much use... would need a similarity-based cutoff rather than a perfect match... not worth it imo
```{r}
ir_inc <- filter(inc, grepl("AthINT", EVENT))

#get cDNA sequences for each IR event
ir_cdna_seq <- data.frame()

for (i in 1:nrow(ir_inc)){
  single_gene_inc <- ir_inc[i, ]
  single_gene_data <- filter(gene_data, model == single_gene_inc$mod & feat == "exon") 
  dir = single_gene_data[1, 5]
  single_gene_cdna <-  select(single_gene_data, start, end)
  single_gene_pos <- select(single_gene_inc, start, end)
  complete_gene_data <- as.data.frame(rbind(single_gene_cdna, single_gene_pos))
  sorted <- complete_gene_data %>% arrange(start)

  seq = data.frame()
  
  for (j in 1:nrow(sorted)){
      chr_num = single_gene_data[1,1]
      chr <- dplyr::filter(athal_gdna, seq.name == as.character(chr_num[1,1]))
      temp <- str_sub(chr$seq.text, sorted[j, 1], sorted[j, 2])
      temp <- paste(temp, collapse = "")
      seq[j, 1] <- temp
      temp <- vector()
  }
  seq <- as.vector(seq$V1)
  seq <- paste(seq, collapse = "")
  if (dir[1, 1] == "-"){
  seq <- as.character(reverseComplement(DNAString(seq)))
  }
  sequence <- data.frame(gene = single_gene_inc$GeneID, event = single_gene_inc$EVENT, sequence = seq)
  ir_cdna_seq <- rbind(ir_cdna_seq, sequence)
}
filter(ir_cdna_seq, gene == "AT1G49740")
```


get all introns from Mao et al (2014). They're alt/cons designations are based on TAIR10, but we can combine all just to get all intron lengths
```{r}
all_introns <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/AS_positions_sequences/Mao_introns.tsv", delim = "\t", col_names = F) 
colnames(all_introns) <- c("chr", "start", "end", "strand", "mod")
all_introns$length <- abs(all_introns$start - all_introns$end)
  
all_introns <- all_introns %>% select(chr, start, end, length) %>% 
  mutate(group = "All introns")
nrow(filter(all_introns, length > 1000))/nrow(all_introns) #<1% introns removed in filtering for graph
```

```{r}
ir_length <- inc %>% filter(grepl("AthINT", EVENT)) %>% 
  mutate(length = abs(as.numeric(start) - as.numeric(end)) + 1) %>% 
  select(chr, start, end, length) %>% 
  mutate(group = "Alternative introns")

nrow(filter(ir_length, length > 1000))/nrow(ir_length) #<3% introns removed by filtering for plot

t.test(ir_length$length, all_introns$length) #p < 2.2e-16, mean alt = 285.6, mean cons = 157.97


lengths <- rbind(all_introns, ir_length)
```

plot intron length
```{r}
png("test.png", units="in", width=5.5, height=3, res=300)
filter(lengths, length <= 1000) %>% 
ggplot(., aes(x = length, y = fct_rev(group), fill = group, colour = group)) + 
  labs(x = "Intron length")+
  theme_minimal()+
  geom_density_ridges(scale = 0.95, alpha = 0.6)+
  scale_fill_manual(values = c("darkblue", "limegreen"))+
  scale_colour_manual(values = c("darkblue", "limegreen"))+
  theme(legend.position = "none",
        axis.title.y = element_blank())+
  geom_signif(y_position = c(1100), xmin = 0.9, xmax = 2.8, annotations = "***", colour = "grey10", tip_length = 0.01, vjust = 0)+
  scale_y_discrete(labels = c("Alternative introns\nn = 598", "All introns\nn = 112 774"))
dev.off()
```

get intron GC contents
start with the alt introns (smaller dataset), then repeat for all introns
```{r}
alt_GC_content <- data.frame()
athal_gdna <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/AS_positions_sequences/Athaliana_167_TAIR10.fa.gz")
athal_gdna$seq.name <- str_sub(athal_gdna$seq.name, 1, 4)
athal_gdna$seq.name <- tolower(athal_gdna$seq.name)

for (i in 1:nrow(ir_length)){
  intron <- ir_length[i, ]
  chr <- filter(athal_gdna, seq.name == as.character(intron[1,1]))
  seq <- str_sub(chr[1, 2], intron[1, 2], intron[1, 3])
  seq <- paste(seq, collapse = "")
  length <- as.numeric(nchar(seq))
  G <- as.numeric(str_count(seq, "G"))
  C <- as.numeric(str_count(seq, "C"))
  GC <- (G+C)/length
  GC_content <- data.frame(GC_content = GC)
  alt_GC_content <- rbind(alt_GC_content, GC_content)
}
alt_GC_content$group <- c("Alternative introns")

#gDNA fasta needs capital letters in seq.name, so go back up and read it back in
athal_gdna <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/AS_positions_sequences/Athaliana_167_TAIR10.fa.gz")
athal_gdna$seq.name <- str_sub(athal_gdna$seq.name, 1, 4)
all_GC_content <- data.frame()

for (i in 1:nrow(all_introns)){
  intron <- all_introns[i, ]
  chr <- filter(athal_gdna, seq.name == as.character(intron[1,1]))
  seq <- str_sub(chr[1, 2], intron[1, 2], intron[1, 3])
  seq <- paste(seq, collapse = "")
  length <- as.numeric(nchar(seq))
  G <- as.numeric(str_count(seq, "G"))
  C <- as.numeric(str_count(seq, "C"))
  GC <- (G+C)/length
  GC_content <- data.frame(GC_content = GC)
  all_GC_content <- rbind(all_GC_content, GC_content)
}
all_GC_content$group <- c("All introns")
t.test(alt_GC_content$GC_content, all_GC_content$GC_content) #not significantly different
ks.test(alt_GC_content$GC_content, all_GC_content$GC_content) #p = 0.005
GC_content <- rbind(alt_GC_content, all_GC_content)

```

plot GC content
```{r}
png("test.png", units="in", width=5.5, height=3, res=300)
ggplot(GC_content, aes(x = GC_content, y = fct_rev(group), fill = group, colour = group)) + 
  labs(x = "Intron GC content")+
  theme_minimal()+
  geom_density_ridges(scale = 0.95, alpha = 0.6)+
  scale_fill_manual(values = c("darkblue", "limegreen"))+
  scale_colour_manual(values = c("darkblue", "limegreen"))+
  theme(legend.position = "none",
        axis.title.y = element_blank())+
  geom_signif(y_position = c(0.9), xmin = 0.9, xmax = 2.8, annotations = "NS", colour = "grey10", tip_length = 0.01, vjust = 0)+
  scale_y_discrete(labels = c("Alternative introns\nn = 598", "All introns\nn = 112 774"))
dev.off()
```

