---
title: "conservation analysis"
output: html_document
---

```{r}
library(phylotools)
library(Biostrings)
library(matrixStats)
library(zoo)
library(tidyverse)
```


```{r}
setwd("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/ortholog_lists/")
files = list.files(pattern="*_orthologs.tsv")
all_orthologs <- do.call(rbind, lapply(files, read_delim, delim = "\t")) %>% 
  select(-c(1, 6))

#dAS_genes includes the 2 non-dAS control genes as well
dAS_genes <- unique(all_orthologs$`Gene Name`)
```

write a function containing a for loop to got through the orthologs and count how many genes a given species has at least 1 ortholog for. this will help with choosing species for the full analysis. 
```{r}
count_orthologs <- function(species){
  count = 0
  for (gene in dAS_genes){
    df <- filter(all_orthologs, `Gene Name` == gene)
    if (species %in% df$`Ortholog Organism Name`){
      count = count + 1
    }
}
  return (count)
}

count_orthologs("Athaliana_TAIR10") #24
#within Arabidopsis
count_orthologs("Alyrata_v2.1") #22
count_orthologs("Ahalleri_v1.1") #22

#within brassicaceae
count_orthologs("BrapaFPsc_v1.3") #23
count_orthologs("Crubella_v1.1") #23

#within eudicots
count_orthologs("Slycopersicum_ITAG4.0") #20
count_orthologs("Stuberosum_v6.1") #19
count_orthologs("Gmax_Wm82.a4.v1") #17
```

```{r}
Alyrata_orthologs <- filter(all_orthologs, `Ortholog Organism Name` == "Alyrata_v2.1")
Ahalleri_orthologs <- filter(all_orthologs, `Ortholog Organism Name` == "Ahalleri_v1.1")
Brapa_orthologs <- filter(all_orthologs, `Ortholog Organism Name` == "BrapaFPsc_v1.3")
Crubella_orthologs <- filter(all_orthologs, `Ortholog Organism Name` == "Crubella_v1.1")
Gmax_orthologs <- filter(all_orthologs, `Ortholog Organism Name` == "Gmax_Wm82.a4.v1")
Slycopersicum_orthologs <- filter(all_orthologs, `Ortholog Organism Name` == "Slycopersicum_ITAG4.0")
Stuberosum_orthologs <- filter(all_orthologs, `Ortholog Organism Name` == "Stuberosum_v6.1")
```


pull out the gene IDs for the sequences we'll want
filter all_orthologs for the correct species name (and version), this will bring with it the Arabidopsis gene ID and the species gene ID. We need the species gene ID to search the transcriptome for the sequence, and we need the Arabidopsis gene ID so we can name the file
repeat for all species (should write a function for this at some point)

Arabidopsis lyrata cDNA
```{r}
Alyrata_transcriptome <- (read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Alyrata.transcript.fa.gz"))
#pull out gene names only
Alyrata_transcriptome$seq.name <- str_sub(Alyrata_transcriptome$seq.name, 1, 9)

for(i in 1:nrow(Alyrata_orthologs)){
  subset <- Alyrata_orthologs[i, ]
  out <- filter(Alyrata_transcriptome, seq.name == subset$`Ortholog Gene Name`)
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Alyrata/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_cDNA.fasta", sep = ""))
}
```

Arabidopsis halleri cDNA
```{r}
Ahalleri_transcriptome <- (read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Ahalleri.transcript.fa.gz"))
#pull out gene names only
Ahalleri_transcriptome$seq.name <- gsub(".[0-9] pacid.*", "", Ahalleri_transcriptome$seq.name)
#Ahalleri_transcriptome$seq.name <- str_sub(Ahalleri_transcriptome$seq.name, 1, 15)

for(i in 1:nrow(Ahalleri_orthologs)){
  subset <- Ahalleri_orthologs[i, ]
  out <- filter(Ahalleri_transcriptome, seq.name == subset$`Ortholog Gene Name`)
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Ahalleri/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_cDNA.fasta", sep = ""))
}
```

Brassica rapa cDNA
```{r}
Brapa_transcriptome <- (read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/BrapaFPsc.transcript.fa.gz"))
#pull out gene names only
Brapa_transcriptome$seq.name <- str_sub(Brapa_transcriptome$seq.name, 1, 12)

for(i in 1:nrow(Brapa_orthologs)){
  subset <- Brapa_orthologs[i, ]
  out <- filter(Brapa_transcriptome, seq.name == subset$`Ortholog Gene Name`)
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Brapa/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_cDNA.fasta", sep = ""))
}
```

Capsella rubella cDNA
```{r}
Crubella_transcriptome <- (read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Crubella.transcript.fa.gz"))
#pull out gene names only
Crubella_transcriptome$seq.name <- str_sub(Crubella_transcriptome$seq.name, 1, 15)

for(i in 1:nrow(Crubella_orthologs)){
  subset <- Crubella_orthologs[i, ]
  out <- filter(Crubella_transcriptome, seq.name == subset$`Ortholog Gene Name`)
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Crubella/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_cDNA.fasta", sep = ""))
}
```

Glycine max cDNA
```{r}
Gmax_transcriptome <- (read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Gmax.transcript.fa.gz"))
#pull out gene names only
Gmax_transcriptome$seq.name <- str_sub(Gmax_transcriptome$seq.name, 1, 15)

for(i in 1:nrow(Gmax_orthologs)){
  subset <- Gmax_orthologs[i, ]
  out <- filter(Gmax_transcriptome, seq.name == subset$`Ortholog Gene Name`)
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Gmax/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_cDNA.fasta", sep = ""))
}
```

Solanum lycopersicum cDNA
```{r}
#tomato-specific workarounds for weird gene IDs
Slycopersicum_orthologs$`Ortholog Gene Name` <- str_replace(Slycopersicum_orthologs$`Ortholog Gene Name`, "gene:", "")
Slycopersicum_orthologs$`Ortholog Gene Name` <- str_sub(Slycopersicum_orthologs$`Ortholog Gene Name`, 1, 14)

Slycopersicum_transcriptome <- (read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Slycopersicum.transcript.fa.gz"))
#pull out gene names only
Slycopersicum_transcriptome$seq.name <- str_sub(Slycopersicum_transcriptome$seq.name, 1, 14)

for(i in 1:nrow(Slycopersicum_orthologs)){
  subset <- Slycopersicum_orthologs[i, ]
  out <- filter(Slycopersicum_transcriptome, seq.name == subset$`Ortholog Gene Name`)
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Slycopersicum/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_cDNA.fasta", sep = ""))
}
```

Solanum tuberosum cDNA
```{r}
Stuberosum_transcriptome <- (read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Stuberosum.transcript.fa.gz"))
#pull out gene names only
Stuberosum_transcriptome$seq.name <- str_sub(Stuberosum_transcriptome$seq.name, 1, 18)

for(i in 1:nrow(Stuberosum_orthologs)){
  subset <- Stuberosum_orthologs[i, ]
  out <- filter(Stuberosum_transcriptome, seq.name == subset$`Ortholog Gene Name`)
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Stuberosum/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_cDNA.fasta", sep = ""))
}
```


for gDNA, need to first read in gff3 and match gene coordinates to ortholog gene IDs. then can pull sequences and write out gDNA fasta files (similar to how cDNA was done)
again I should write a function for this 

Arabidopsis lyrata gDNA
```{r}
Alyrata_gff <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Alyrata.gene.gff3", delim = "\t", skip = 3, col_names = F) %>% 
  filter(X3 == "gene") %>% 
  mutate(X9 = str_sub(X9, 4, 12)) %>% 
  select(c(X1, X4, X5, X7, X9))
colnames(Alyrata_gff) <- c("scaffold", "start", "end", "strand", "Ortholog Gene Name")
Alyrata_orthologs <- left_join(Alyrata_orthologs, Alyrata_gff, by = "Ortholog Gene Name")

Alyrata_genome <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Alyrata.fa.gz") %>% 
  filter(seq.name %in% Alyrata_orthologs$scaffold)

for(i in 1:nrow(Alyrata_orthologs)){
  subset <- Alyrata_orthologs[i, ]
  scaffold <- filter(Alyrata_genome, seq.name == subset$scaffold)
  scaffold$seq.text <- str_sub(scaffold$seq.text, subset$start, subset$end )
  if (subset$strand == "-"){
    scaffold$seq.text <- as.character(reverseComplement(DNAString(scaffold$seq.text)))
  }
  scaffold$seq.name <- paste(subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA", sep = "")
  dat2fasta(scaffold, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Alyrata/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA.fasta", sep = ""))
}
```

Arabidopsis halleri gDNA
```{r}
Ahalleri_gff <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Ahalleri.gene.gff3", delim = "\t", skip = 3, col_names = F) %>% 
  filter(X3 == "gene")%>% 
  mutate(X9 = gsub(".v1.1.*", "", X9)) %>% 
  mutate(X9 = gsub("ID=", "", X9)) %>% 
  select(c(X1, X4, X5, X7, X9))
colnames(Ahalleri_gff) <- c("scaffold", "start", "end", "strand", "Ortholog Gene Name")
Ahalleri_orthologs <- left_join(Ahalleri_orthologs, Ahalleri_gff, by = "Ortholog Gene Name")

Ahalleri_genome <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Ahalleri.fa.gz") %>% 
  filter(seq.name %in% Ahalleri_orthologs$scaffold)

for(i in 1:nrow(Ahalleri_orthologs)){
  subset <- Ahalleri_orthologs[i, ]
  scaffold <- filter(Ahalleri_genome, seq.name == subset$scaffold)
  scaffold$seq.text <- str_sub(scaffold$seq.text, subset$start, subset$end )
  if (subset$strand == "-"){
    scaffold$seq.text <- as.character(reverseComplement(DNAString(scaffold$seq.text)))
  }
    scaffold$seq.name <- paste(subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA", sep = "")
  dat2fasta(scaffold, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Ahalleri/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA.fasta", sep = ""))
}
```

Brassica rapa gDNA
```{r}
Brapa_gff <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/BrapaFPSc.gene.gff3", delim = "\t", skip = 3, col_names = F) %>% 
  filter(X3 == "gene")%>% 
  mutate(X9 = gsub(".v1.3.*", "", X9)) %>% 
  mutate(X9 = gsub("ID=", "", X9)) %>% 
  select(c(X1, X4, X5, X7, X9))
colnames(Brapa_gff) <- c("scaffold", "start", "end", "strand", "Ortholog Gene Name")
Brapa_orthologs <- left_join(Brapa_orthologs, Brapa_gff, by = "Ortholog Gene Name")

Brapa_genome <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/BrapaFPSc.fa.gz") %>% 
  filter(seq.name %in% Brapa_orthologs$scaffold)

for(i in 1:nrow(Brapa_orthologs)){
  subset <- Brapa_orthologs[i, ]
  scaffold <- filter(Brapa_genome, seq.name == subset$scaffold)
  scaffold$seq.text <- str_sub(scaffold$seq.text, subset$start, subset$end )
  if (subset$strand == "-"){
    scaffold$seq.text <- as.character(reverseComplement(DNAString(scaffold$seq.text)))
  }
  scaffold$seq.name <- paste(subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA", sep = "")
  dat2fasta(scaffold, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Brapa/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA.fasta", sep = ""))
}
```

Capsella rubella gDNA
```{r}
Crubella_gff <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Crubella.gene.gff3", delim = "\t", skip = 3, col_names = F) %>% 
  filter(X3 == "gene")%>% 
  mutate(X9 = gsub(".v1.1.*", "", X9)) %>% 
  mutate(X9 = gsub("ID=", "", X9)) %>% 
  select(c(X1, X4, X5, X7, X9))
colnames(Crubella_gff) <- c("scaffold", "start", "end", "strand", "Ortholog Gene Name")
Crubella_orthologs <- left_join(Crubella_orthologs, Crubella_gff, by = "Ortholog Gene Name")

Crubella_genome <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Crubella.fa.gz") %>% 
  filter(seq.name %in% Crubella_orthologs$scaffold)

for(i in 1:nrow(Crubella_orthologs)){
  subset <- Crubella_orthologs[i, ]
  scaffold <- filter(Crubella_genome, seq.name == subset$scaffold)
  scaffold$seq.text <- str_sub(scaffold$seq.text, subset$start, subset$end )
  if (subset$strand == "-"){
    scaffold$seq.text <- as.character(reverseComplement(DNAString(scaffold$seq.text)))
  }
  scaffold$seq.name <- paste(subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA", sep = "")
  dat2fasta(scaffold, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Crubella/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA.fasta", sep = ""))
}
```

Glycine max gDNA
```{r}
Gmax_gff <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Gmax.gene.gff3", delim = "\t", skip = 3, col_names = F) %>% 
  filter(X3 == "gene")%>% 
  mutate(X9 = gsub(".Wm82.*", "", X9)) %>% 
  mutate(X9 = gsub("ID=", "", X9)) %>% 
  select(c(X1, X4, X5, X7, X9))
colnames(Gmax_gff) <- c("scaffold", "start", "end", "strand", "Ortholog Gene Name")
Gmax_orthologs <- left_join(Gmax_orthologs, Gmax_gff, by = "Ortholog Gene Name")

Gmax_genome <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Gmax.fa.gz") %>% 
  filter(seq.name %in% Gmax_orthologs$scaffold)

for(i in 1:nrow(Gmax_orthologs)){
  subset <- Gmax_orthologs[i, ]
  scaffold <- filter(Gmax_genome, seq.name == subset$scaffold)
  scaffold$seq.text <- str_sub(scaffold$seq.text, subset$start, subset$end)
  if (subset$strand == "-"){
    scaffold$seq.text <- as.character(reverseComplement(DNAString(scaffold$seq.text)))
  }
  scaffold$seq.name <- paste(subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA", sep = "")  
  dat2fasta(scaffold, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Gmax/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA.fasta", sep = ""))
}
```

Solanum lycopersicum gDNA
```{r}
#tomato-specific workarounds for weird gene IDs
Slycopersicum_orthologs$`Ortholog Gene Name` <- str_replace(Slycopersicum_orthologs$`Ortholog Gene Name`, "gene:", "")
Slycopersicum_orthologs$`Ortholog Gene Name` <- str_sub(Slycopersicum_orthologs$`Ortholog Gene Name`, 1, 14)

Slycopersicum_gff <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Slycopersicum.gene.gff3", delim = "\t", skip = 6, col_names = F)%>% 
  filter(X3 == "gene")%>% 
  mutate(X9 = gsub(".ITAG.*", "", X9)) %>% 
  mutate(X9 = gsub("ID=", "", X9)) %>% 
  select(c(X1, X4, X5, X7, X9))
colnames(Slycopersicum_gff) <- c("scaffold", "start", "end", "strand", "Ortholog Gene Name")

#tomato-specific workarounds for weird gene IDs
Slycopersicum_gff$`Ortholog Gene Name` <- str_replace(Slycopersicum_gff$`Ortholog Gene Name`, "gene:", "")
Slycopersicum_gff$`Ortholog Gene Name` <- str_sub(Slycopersicum_gff$`Ortholog Gene Name`, 1, 14)

Slycopersicum_orthologs <- left_join(Slycopersicum_orthologs, Slycopersicum_gff, by = "Ortholog Gene Name")

Slycopersicum_genome <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Slycopersicum.fa.gz")

#tomato-specific idiocy from spaces at the end of sequence names
Slycopersicum_genome$seq.name <- str_sub(Slycopersicum_genome$seq.name, 1, 9)

for(i in 1:nrow(Slycopersicum_orthologs)){
  subset <- Slycopersicum_orthologs[i, ]
  scaffold <- filter(Slycopersicum_genome, seq.name == subset$scaffold)
  scaffold$seq.text <- str_sub(scaffold$seq.text, subset$start, subset$end)
  if (subset$strand == "-"){
    scaffold$seq.text <- as.character(reverseComplement(DNAString(scaffold$seq.text)))
  }
  scaffold$seq.name <- paste(subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA", sep = "")  
  dat2fasta(scaffold, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Slycopersicum/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA.fasta", sep = ""))
}
```

Solanum tuberosum gDNA
```{r}
Stuberosum_gff <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Stuberosum.gene.gff3", delim = "\t", skip = 6, col_names = F) %>% 
  filter(X3 == "gene")%>% 
  mutate(X9 = gsub(".v6.1.*", "", X9)) %>% 
  mutate(X9 = gsub("ID=", "", X9)) %>% 
  select(c(X1, X4, X5, X7, X9))
colnames(Stuberosum_gff) <- c("scaffold", "start", "end", "strand", "Ortholog Gene Name")
Stuberosum_orthologs <- left_join(Stuberosum_orthologs, Stuberosum_gff, by = "Ortholog Gene Name")

Stuberosum_genome <- read.fasta("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_files/Stuberosum.fa.gz") %>% 
  filter(seq.name %in% Stuberosum_orthologs$scaffold)

for(i in 1:nrow(Stuberosum_orthologs)){
  subset <- Stuberosum_orthologs[i, ]
  scaffold <- filter(Stuberosum_genome, seq.name == subset$scaffold)
  scaffold$seq.text <- str_sub(scaffold$seq.text, subset$start, subset$end)
  if (subset$strand == "-"){
    scaffold$seq.text <- as.character(reverseComplement(DNAString(scaffold$seq.text)))
  }
  scaffold$seq.name <- paste(subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA", sep = "")  
  dat2fasta(scaffold, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/species_sequences/Stuberosum/",subset$`Gene Name`, "_", subset$`Ortholog Gene Name`, "_gDNA.fasta", sep = ""))
}
```

since 2 dAS genes have only 1 intron, we can't select constitutive introns from these genes, so 2 additional IR events need to be selected from the vast DB. randomly select number between 273329 and 388522 (IR events in vast DB), check to make sure it has sufficient read depth across K20d and K20w. 
```{r}
set.seed(44)
sample(273329:388522, 1)
#291867: PUB33(AT2G45910)	AthINT0096724	(IR4) chr2:18895250-18895386

set.seed(49)
sample(273329:388522, 1)
#305301: AT2G41190 AthINT0020254 (IR7)	chr2:17168244-17168343
```

after downloading ortholog lists for the new genes, re-run the code above to get cDNA and gDNA fasta files (this will re-write the files for the genes we had before, but it doesn't matter)


didn't end up using this, phyloFit model built using a single gDNA sequence and will be adjusted to each gene within phastCons.
```{r}
#create a single fasta file for each species containing the gDNA sequences for the 25 genes of interest (more or less depending on if orthologs are missing, or if genes are duplicates). Will do A thaliana separately (likely manually) as those sequences are arranged differently
species <- c("Alyrata", "Ahalleri", "Brapa", "Crubella", "Gmax", "Slycopersicum", "Stuberosum")
for (plant in species){
  gDNA = list.files(path = paste("species_sequences/", plant, "/", sep = ""), pattern="*_gDNA.fasta", full.names = T)
  gDNA <- do.call(rbind, lapply(gDNA, read.fasta))
  gDNA$seq.name <- paste(plant, ":", gDNA$seq.name, ":1:+:", nchar(gDNA$seq.text), sep = "")
  dat2fasta(gDNA, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/TBA/gDNA/", plant, ".fasta", sep = ""))
}
```

for all the flanking exon and alt intron sequences we pulled, need to get sequences of the same length. From A thaliana, the shortest exon is 48nt and the shortest intron in 76nt. based on these values, a sequence length of 40nt makes sense, as this is nearly as long as the shortest exon, and will almost split in half the shortest intron. We'll have 4 types of sequencse to analyze: 
1. 40 nt directly upstream of intron (last 40nt of upstream exon)
2. 40 nt directly downstream of intron (first 40 nt of downstream intron)
3. 40 nt at 5' end of intron (first 40 nt of intron)
4. 40 nt at 3' end of intron (last 40 nt of intron)
Need to make a fasta file for each of these 4 sequence types and for each gene. Thinking of a subdirectory for each sequence type containing all the single gene files (which will make later analysis easier)
```{r}
species <- c("Athaliana", "Alyrata", "Ahalleri", "Brapa", "Crubella", "Gmax", "Slycopersicum", "Stuberosum")
alt_genes <- dAS_genes[! dAS_genes %in% c("AT2G41190", "AT2G45910")]

#upstream exons
upstream_exons <- data.frame()
for (plant in species){
  seq = list.files(path = paste("conserved_sequences/alternative/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*_upstream_exon", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, start = -40)
  upstream_exons <- rbind(upstream_exons, seq)
}

for (gene in alt_genes){
  out <- upstream_exons %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata",
    str_detect(seq.name, "Brara") ~ "Brapa",
    str_detect(seq.name, "Carub") ~ "Crubella",
    str_detect(seq.name, "Glyma") ~ "Gmax",
    str_detect(seq.name, "Soltu") ~ "Stuberosum",
    str_detect(seq.name, "Solyc") ~ "Slycopersicum"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/alternative/upstream_exon/", gene, ".fasta", sep = ""))
}

#downstream_exons
downstream_exons <- data.frame()
for (plant in species){
  seq = list.files(path = paste("conserved_sequences/alternative/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*_downstream_exon", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, end = 40)
  downstream_exons <- rbind(downstream_exons, seq)
}

for (gene in alt_genes){
  out <- downstream_exons %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata",
    str_detect(seq.name, "Brara") ~ "Brapa",
    str_detect(seq.name, "Carub") ~ "Crubella",
    str_detect(seq.name, "Glyma") ~ "Gmax",
    str_detect(seq.name, "Soltu") ~ "Stuberosum",
    str_detect(seq.name, "Solyc") ~ "Slycopersicum"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/alternative/downstream_exon/", gene, ".fasta", sep = ""))
}

#5' end of intron
intron5 <- data.frame()
for (plant in species){
  seq = list.files(path = paste("conserved_sequences/alternative/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*intron", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, end = 40)
  intron5 <- rbind(intron5, seq)
}

for (gene in alt_genes){
  out <- intron5 %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata",
    str_detect(seq.name, "Brara") ~ "Brapa",
    str_detect(seq.name, "Carub") ~ "Crubella",
    str_detect(seq.name, "Glyma") ~ "Gmax",
    str_detect(seq.name, "Soltu") ~ "Stuberosum",
    str_detect(seq.name, "Solyc") ~ "Slycopersicum"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/alternative/intron5/", gene, ".fasta", sep = ""))
}

#3' end of intron
intron3 <- data.frame()
for (plant in species){
  seq = list.files(path = paste("conserved_sequences/alternative/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*intron", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, start = -40)
  intron3 <- rbind(intron3, seq)
}

for (gene in alt_genes){
  out <- intron3 %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata",
    str_detect(seq.name, "Brara") ~ "Brapa",
    str_detect(seq.name, "Carub") ~ "Crubella",
    str_detect(seq.name, "Glyma") ~ "Gmax",
    str_detect(seq.name, "Soltu") ~ "Stuberosum",
    str_detect(seq.name, "Solyc") ~ "Slycopersicum"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  #out <- out %>% mutate (seq.name = replace_na("Athaliana"))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/alternative/intron3/", gene, ".fasta", sep = ""))
}
```

read back in the phastcons scores for each sequence type and plot as a function of position
upstream exons
```{r}
upexon_scores = list.files(path = "phastCons/alternative/upstream_exon/", pattern = "fasta.wig", full.names = T)
upexon_scores <- do.call(cbind, lapply(upexon_scores, read_delim, delim = "\n"))
colnames(upexon_scores) <- alt_genes
upexon_scores <- upexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
upexon_scores$med <- rowMedians(as.matrix(upexon_scores[, c(1:23)]))
median <- upexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
upexon_scores <- cbind(upexon_scores, median)
#this is ugly, but it'll do :)

upexon_scores$position <- c(1:40)
ggplot(upexon_scores, aes(x = position, y = median))+
  geom_line()
```

downstream exons
```{r}
downexon_scores = list.files(path = "phastCons/alternative/downstream_exon/", pattern = ".wig", full.names = T)
downexon_scores <- do.call(cbind, lapply(downexon_scores, read_delim, delim = "\n"))
colnames(downexon_scores) <- alt_genes
downexon_scores <- downexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
downexon_scores$med <- rowMedians(as.matrix(downexon_scores[, c(1:23)]))
median <- downexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
downexon_scores <- cbind(downexon_scores, median)
#this is ugly, but it'll do :)

downexon_scores$position <- c(1:40)
ggplot(downexon_scores, aes(x = position, y = median))+
  geom_line()
```

intron 5'
```{r}
intron5_scores = list.files(path = "phastCons/alternative/intron5/", pattern = ".wig", full.names = T)
intron5_scores <- do.call(cbind, lapply(intron5_scores, read_delim, delim = "\n"))
colnames(intron5_scores) <- alt_genes
intron5_scores <- intron5_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron5_scores$med <- rowMedians(as.matrix(intron5_scores[, c(1:23)]))
median <- intron5_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
intron5_scores <- cbind(intron5_scores, median)
#this is ugly, but it'll do :)

intron5_scores$position <- c(1:40)
ggplot(intron5_scores, aes(x = position, y = median))+
  geom_line()
```

intron 3'
```{r}
intron3_scores = list.files(path = "phastCons/alternative/intron3/", pattern = ".wig", full.names = T)
intron3_scores <- do.call(cbind, lapply(intron3_scores, read_delim, delim = "\n"))
colnames(intron3_scores) <- alt_genes
intron3_scores <- intron3_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron3_scores$med <- rowMedians(as.matrix(intron3_scores[, c(1:23)]))
median <- intron3_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
intron3_scores <- cbind(intron3_scores, median)
#this is ugly, but it'll do :)

intron3_scores$position <- c(1:40)
ggplot(intron3_scores, aes(x = position, y = median))+
  geom_line()
```

don't like how these look, think it's an issue of the neutral models. let's instead build a neutral model for each gene using all its gDNA sequences. write out all gDNA from the same gene to a fasta file, then feed it into mafft. in cases where a single species has multiple ortholgs, select one at random (do this after, manually. see selected genes in conservation.xlsx).
```{r}
species <- c("Athaliana", "Alyrata", "Ahalleri", "Brapa", "Crubella", "Gmax", "Slycopersicum", "Stuberosum")
gDNA <- data.frame()
for (plant in species){
  seq = list.files(path = paste("species_sequences/", plant, "/", sep = ""), pattern = "_gDNA.fasta", full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  gDNA <- rbind(gDNA, seq)
}

for (gene in dAS_genes){
  out <- gDNA %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata",
    str_detect(seq.name, "Brara") ~ "Brapa",
    str_detect(seq.name, "Carub") ~ "Crubella",
    str_detect(seq.name, "Glyma") ~ "Gmax",
    str_detect(seq.name, "Soltu") ~ "Stuberosum",
    str_detect(seq.name, "Solyc") ~ "Slycopersicum"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/models/", gene, "_unaligned.fasta", sep = ""))
}
```

read in the phyloP scores and plot
upstream_exon
```{r}
upexon_scores = list.files(path = "phastCons/alternative/upstream_exon/", pattern = "_phylop.wig", full.names = T)
upexon_scores <- do.call(cbind, lapply(upexon_scores, read_delim, delim = "\n"))
colnames(upexon_scores) <- alt_genes
upexon_scores <- upexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
upexon_scores$med <- rowMedians(as.matrix(upexon_scores[, c(1:23)]))
median <- upexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
upexon_scores <- cbind(upexon_scores, median)
#this is ugly, but it'll do :)

upexon_scores$position <- c(-40:-1)
ggplot(upexon_scores, aes(x = position, y = median))+
  geom_line()

#try means instead?
mean <- upexon_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
upexon_scores <- cbind(upexon_scores, mean)

upexon_scores$position <- c(-40:-1)
ggplot(upexon_scores, aes(x = position, y = mean))+
  geom_line()
```

downstream exons
```{r}
downexon_scores = list.files(path = "phastCons/alternative/downstream_exon/", pattern = "_phylop.wig", full.names = T)
downexon_scores <- do.call(cbind, lapply(downexon_scores, read_delim, delim = "\n"))
colnames(downexon_scores) <- alt_genes
downexon_scores <- downexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
downexon_scores$med <- rowMedians(as.matrix(downexon_scores[, c(1:23)]))
median <- downexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
downexon_scores <- cbind(downexon_scores, median)
#this is ugly, but it'll do :)

downexon_scores$position <- c(1:40)
ggplot(downexon_scores, aes(x = position, y = median))+
  geom_line()

#try means instead?
mean <- downexon_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
downexon_scores <- cbind(downexon_scores, mean)
```

intron 5'
```{r}
intron5_scores = list.files(path = "phastCons/alternative/intron5/", pattern = "_phylop.wig", full.names = T)
intron5_scores <- do.call(cbind, lapply(intron5_scores, read_delim, delim = "\n"))
colnames(intron5_scores) <- alt_genes
intron5_scores <- intron5_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron5_scores$med <- rowMedians(as.matrix(intron5_scores[, c(1:23)]))
median <- intron5_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
intron5_scores <- cbind(intron5_scores, median)
#this is ugly, but it'll do :)

intron5_scores$position <- c(1:40)
ggplot(intron5_scores, aes(x = position, y = median))+
  geom_line()

#try means instead?
mean <- intron5_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
intron5_scores <- cbind(intron5_scores, mean)
```

intron 3'
```{r}
intron3_scores = list.files(path = "phastCons/alternative/intron3/", pattern = "_phylop.wig", full.names = T)
intron3_scores <- do.call(cbind, lapply(intron3_scores, read_delim, delim = "\n"))
colnames(intron3_scores) <- alt_genes
intron3_scores <- intron3_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron3_scores$med <- rowMedians(as.matrix(intron3_scores[, c(1:23)]))
median <- intron3_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
intron3_scores <- cbind(intron3_scores, median)
#this is ugly, but it'll do :)

intron3_scores$position <- c(-40:-1)
ggplot(intron3_scores, aes(x = position, y = median))+
  geom_line()+
  geom_hline(yintercept = 0)

#try means instead?
mean <- intron3_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
intron3_scores <- cbind(intron3_scores, mean)
```

these look great for the introns, not so much the exons. gonna try re-running phastcons using the gene-specific gDNA neutral models.

read in the phastCons scores and plot
upstream_exon
```{r}
upexon_scores = list.files(path = "phastCons/alternative/upstream_exon/", pattern = "_phastcons.wig", full.names = T)
upexon_scores <- do.call(cbind, lapply(upexon_scores, read_delim, delim = "\n"))
colnames(upexon_scores) <- alt_genes
upexon_scores <- upexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
upexon_scores$med <- rowMedians(as.matrix(upexon_scores[, c(1:23)]))
median <- upexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
upexon_scores <- cbind(upexon_scores, median)
#this is ugly, but it'll do :)

upexon_scores$position <- c(1:40)
ggplot(upexon_scores, aes(x = position, y = median))+
  geom_line()
```

downstream exons
```{r}
downexon_scores = list.files(path = "phastCons/alternative/downstream_exon/", pattern = "_phastcons.wig", full.names = T)
downexon_scores <- do.call(cbind, lapply(downexon_scores, read_delim, delim = "\n"))
colnames(downexon_scores) <- alt_genes
downexon_scores <- downexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
downexon_scores$med <- rowMedians(as.matrix(downexon_scores[, c(1:23)]))
median <- downexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
downexon_scores <- cbind(downexon_scores, median)
#this is ugly, but it'll do :)

downexon_scores$position <- c(1:40)
ggplot(downexon_scores, aes(x = position, y = median))+
  geom_line()
```

intron 5'
```{r}
intron5_scores = list.files(path = "phastCons/alternative/intron5/", pattern = "_phastcons.wig", full.names = T)
intron5_scores <- do.call(cbind, lapply(intron5_scores, read_delim, delim = "\n"))
colnames(intron5_scores) <- alt_genes
intron5_scores <- intron5_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron5_scores$med <- rowMedians(as.matrix(intron5_scores[, c(1:23)]))
median <- intron5_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
intron5_scores <- cbind(intron5_scores, median)
#this is ugly, but it'll do :)

intron5_scores$position <- c(1:40)
ggplot(intron5_scores, aes(x = position, y = median))+
  geom_line()
```

intron 3'
```{r}
intron3_scores = list.files(path = "phastCons/alternative/intron3/", pattern = "_phastcons.wig", full.names = T)
intron3_scores <- do.call(cbind, lapply(intron3_scores, read_delim, delim = "\n"))
colnames(intron3_scores) <- alt_genes
intron3_scores <- intron3_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron3_scores$med <- rowMedians(as.matrix(intron3_scores[, c(1:23)]))
median <- intron3_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
intron3_scores <- cbind(intron3_scores, median)
#this is ugly, but it'll do :)

intron3_scores$position <- c(1:40)
ggplot(intron3_scores, aes(x = position, y = median))+
  geom_line()
```

moving on to constitutive introns: after mafft alignment, read in the sequences and write them out according to gene and type 
```{r}
species <- c("Athaliana", "Alyrata", "Ahalleri", "Brapa", "Crubella", "Gmax", "Slycopersicum", "Stuberosum")
cons_genes <- dAS_genes[! dAS_genes %in% c("AT1G24147", "AT1G53840")]
cons_genes_struct <- cons_genes[! cons_genes %in% "AT2G47730"]

#upstream exons
upstream_exons <- data.frame()
for (plant in species){
  seq = list.files(path = paste("conserved_sequences/constitutive/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*_upstream_exon", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, start = -40)
  upstream_exons <- rbind(upstream_exons, seq)
}

for (gene in cons_genes){
  out <- upstream_exons %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata",
    str_detect(seq.name, "Brara") ~ "Brapa",
    str_detect(seq.name, "Carub") ~ "Crubella",
    str_detect(seq.name, "Glyma") ~ "Gmax",
    str_detect(seq.name, "Soltu") ~ "Stuberosum",
    str_detect(seq.name, "Solyc") ~ "Slycopersicum"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/constitutive/upstream_exon/", gene, ".fasta", sep = ""))
}

#downstream_exons
downstream_exons <- data.frame()
for (plant in species){
  seq = list.files(path = paste("conserved_sequences/constitutive/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*_downstream_exon", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, end = 40)
  downstream_exons <- rbind(downstream_exons, seq)
}

for (gene in cons_genes){
  out <- downstream_exons %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata",
    str_detect(seq.name, "Brara") ~ "Brapa",
    str_detect(seq.name, "Carub") ~ "Crubella",
    str_detect(seq.name, "Glyma") ~ "Gmax",
    str_detect(seq.name, "Soltu") ~ "Stuberosum",
    str_detect(seq.name, "Solyc") ~ "Slycopersicum"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/constitutive/downstream_exon/", gene, ".fasta", sep = ""))
}

#5' end of intron
intron5 <- data.frame()
for (plant in species){
  seq = list.files(path = paste("conserved_sequences/constitutive/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*intron", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, end = 40)
  intron5 <- rbind(intron5, seq)
}

for (gene in cons_genes){
  out <- intron5 %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata",
    str_detect(seq.name, "Brara") ~ "Brapa",
    str_detect(seq.name, "Carub") ~ "Crubella",
    str_detect(seq.name, "Glyma") ~ "Gmax",
    str_detect(seq.name, "Soltu") ~ "Stuberosum",
    str_detect(seq.name, "Solyc") ~ "Slycopersicum"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/constitutive/intron5/", gene, ".fasta", sep = ""))
}

#3' end of intron
intron3 <- data.frame()
for (plant in species){
  seq = list.files(path = paste("conserved_sequences/constitutive/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*intron", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, start = -40)
  intron3 <- rbind(intron3, seq)
}

for (gene in cons_genes){
  out <- intron3 %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata",
    str_detect(seq.name, "Brara") ~ "Brapa",
    str_detect(seq.name, "Carub") ~ "Crubella",
    str_detect(seq.name, "Glyma") ~ "Gmax",
    str_detect(seq.name, "Soltu") ~ "Stuberosum",
    str_detect(seq.name, "Solyc") ~ "Slycopersicum"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  #out <- out %>% mutate (seq.name = replace_na("Athaliana"))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/constitutive/intron3/", gene, ".fasta", sep = ""))
}
```

read in the phyloP scores for constitutive introns and plot alongside alternatives
upstream_exon
```{r}
upexon_scores_c = list.files(path = "phastCons/constitutive/upstream_exon/", pattern = "_phylop.wig", full.names = T)
upexon_scores_c <- do.call(cbind, lapply(upexon_scores_c, read_delim, delim = "\n"))
colnames(upexon_scores_c) <- cons_genes_struct
upexon_scores_c <- upexon_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
upexon_scores_c$med <- rowMedians(as.matrix(upexon_scores_c[, c(1:23)]))
median_c <- upexon_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
upexon_scores_c <- cbind(upexon_scores_c, median_c)
#this is ugly, but it'll do :)

#try means instead?
mean <- upexon_scores_c$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
upexon_scores_c <- cbind(upexon_scores_c, mean)

upexon_scores_plot <- as.data.frame(cbind(upexon_scores$position, upexon_scores$median, upexon_scores_c$median_c, upexon_scores$mean, mean))
colnames(upexon_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
upexon_scores_plot <-  upexon_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(upexon_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  labs(title = "Upstream exons - all species",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 5' SS")+
  theme(legend.title = element_blank())
```

downstream exons
```{r}
downexon_scores_c = list.files(path = "phastCons/constitutive/downstream_exon/", pattern = "_phylop.wig", full.names = T)
downexon_scores_c <- do.call(cbind, lapply(downexon_scores_c, read_delim, delim = "\n"))
colnames(downexon_scores_c) <- cons_genes_struct
downexon_scores_c <- downexon_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
downexon_scores_c$med <- rowMedians(as.matrix(downexon_scores_c[, c(1:23)]))
median_c <- downexon_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
downexon_scores_c <- cbind(downexon_scores_c, median_c)
#this is ugly, but it'll do :)

#try means instead?
mean <- downexon_scores_c$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
downexon_scores_c <- cbind(downexon_scores_c, mean)

downexon_scores_plot <- as.data.frame(cbind(downexon_scores$position, downexon_scores$median, downexon_scores_c$median_c, downexon_scores$mean, mean))
colnames(downexon_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
downexon_scores_plot <-  downexon_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(downexon_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  labs(title = "Downstream exons - all species",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 3' SS")+
  theme(legend.title = element_blank())
```

intron 5'
```{r}
intron5_c = list.files(path = "phastCons/constitutive/intron5/", pattern = "_phylop.wig", full.names = T)
intron5_c <- do.call(cbind, lapply(intron5_c, read_delim, delim = "\n"))
colnames(intron5_c) <- cons_genes_struct
intron5_c <- intron5_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron5_c$med <- rowMedians(as.matrix(intron5_c[, c(1:23)]))
median_c <- intron5_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
intron5_c <- cbind(intron5_c, median_c)
#this is ugly, but it'll do :)

#try means instead?
mean <- intron5_c$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
intron5_c <- cbind(intron5_c, mean)

intron5_scores_plot <- as.data.frame(cbind(intron5_scores$position, intron5_scores$median, intron5_c$median_c, intron5_scores$mean, mean))
colnames(intron5_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
intron5_scores_plot <-  intron5_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(intron5_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  labs(title = "Intron 5' end - all species",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 5' SS")+
  theme(legend.title = element_blank())
```

intron 3'
```{r}
intron3_c = list.files(path = "phastCons/constitutive/intron3/", pattern = "_phylop.wig", full.names = T)
intron3_c <- do.call(cbind, lapply(intron3_c, read_delim, delim = "\n"))
colnames(intron3_c) <- cons_genes_struct
intron3_c <- intron3_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron3_c$med <- rowMedians(as.matrix(intron3_c[, c(1:23)]))
median_c <- intron3_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
intron3_c <- cbind(intron3_c, median_c)
#this is ugly, but it'll do :)

#try means instead?
mean <- intron3_c$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
intron3_c <- cbind(intron3_c, mean)

intron3_scores_plot <- as.data.frame(cbind(intron3_scores$position, intron3_scores$median, intron3_c$median_c, intron3_scores$mean, mean))
colnames(intron3_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
intron3_scores_plot <-  intron3_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(intron3_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  labs(title = "Intron 3' end - all species",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 3' SS")+
  theme(legend.title = element_blank())
```

what if we only look within Arabidopsis species? set up the phyloP input fasta files using only Arabidopsis sequences
```{r}
arabidopsis <- c("Athaliana", "Alyrata", "Ahalleri")
cons_genes <- dAS_genes[! dAS_genes %in% c("AT1G24147", "AT1G53840")]
cons_genes_struct <- cons_genes[! cons_genes %in% "AT2G47730"]

#alternative
#upstream exons
upstream_exons <- data.frame()
for (plant in arabidopsis){
  seq = list.files(path = paste("conserved_sequences/alternative/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*_upstream_exon", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, start = -40)
  upstream_exons <- rbind(upstream_exons, seq)
}

for (gene in cons_genes){
  out <- upstream_exons %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/Arabidopsis/alternative/upstream_exon/", gene, ".fasta", sep = ""))
}

#downstream_exons
downstream_exons <- data.frame()
for (plant in arabidopsis){
  seq = list.files(path = paste("conserved_sequences/alternative/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*_downstream_exon", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, end = 40)
  downstream_exons <- rbind(downstream_exons, seq)
}

for (gene in cons_genes){
  out <- downstream_exons %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/Arabidopsis/alternative/downstream_exon/", gene, ".fasta", sep = ""))
}

#5' end of intron
intron5 <- data.frame()
for (plant in arabidopsis){
  seq = list.files(path = paste("conserved_sequences/alternative/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*intron", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, end = 40)
  intron5 <- rbind(intron5, seq)
}

for (gene in cons_genes){
  out <- intron5 %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/Arabidopsis/alternative/intron5/", gene, ".fasta", sep = ""))
}

#3' end of intron
intron3 <- data.frame()
for (plant in arabidopsis){
  seq = list.files(path = paste("conserved_sequences/alternative/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*intron", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, start = -40)
  intron3 <- rbind(intron3, seq)
}

for (gene in cons_genes){
  out <- intron3 %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  #out <- out %>% mutate (seq.name = replace_na("Athaliana"))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/Arabidopsis/alternative/intron3/", gene, ".fasta", sep = ""))
}

#constitutive
#upstream exons
upstream_exons <- data.frame()
for (plant in arabidopsis){
  seq = list.files(path = paste("conserved_sequences/constitutive/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*_upstream_exon", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, start = -40)
  upstream_exons <- rbind(upstream_exons, seq)
}

for (gene in cons_genes){
  out <- upstream_exons %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/Arabidopsis/constitutive/upstream_exon/", gene, ".fasta", sep = ""))
}

#downstream_exons
downstream_exons <- data.frame()
for (plant in arabidopsis){
  seq = list.files(path = paste("conserved_sequences/constitutive/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*_downstream_exon", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, end = 40)
  downstream_exons <- rbind(downstream_exons, seq)
}

for (gene in cons_genes){
  out <- downstream_exons %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/Arabidopsis/constitutive/downstream_exon/", gene, ".fasta", sep = ""))
}

#5' end of intron
intron5 <- data.frame()
for (plant in arabidopsis){
  seq = list.files(path = paste("conserved_sequences/constitutive/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*intron", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, end = 40)
  intron5 <- rbind(intron5, seq)
}

for (gene in cons_genes){
  out <- intron5 %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/Arabidopsis/constitutive/intron5/", gene, ".fasta", sep = ""))
}

#3' end of intron
intron3 <- data.frame()
for (plant in arabidopsis){
  seq = list.files(path = paste("conserved_sequences/constitutive/", plant, "/", sep = ""), full.names = T)
  seq <- do.call(rbind, lapply(seq, read.fasta))
  seq <- filter(seq, grepl("*intron", seq.name))
  seq$seq.text <- str_sub(seq$seq.text, start = -40)
  intron3 <- rbind(intron3, seq)
}

for (gene in cons_genes){
  out <- intron3 %>% filter(str_detect(seq.name, as.character(gene)))
  out <- out %>% mutate(seq.name = case_when(
    str_detect(seq.name, "Araha") ~ "Ahalleri",
    str_detect(seq.name, "AL") ~ "Alyrata"
  )) %>% 
    mutate(seq.name = ifelse(is.na(seq.name), "Athaliana", seq.name))
  #out <- out %>% mutate (seq.name = replace_na("Athaliana"))
  dat2fasta(out, outfile = paste("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phastCons/Arabidopsis/constitutive/intron3/", gene, ".fasta", sep = ""))
}
```

upstream exons
```{r}
alt_genes_ara <- alt_genes[! alt_genes %in% c("AT1G24147", "AT1G53840")]
cons_genes_ara <- cons_genes_struct[! cons_genes_struct %in% "AT5G15960"]
upexon_scores = list.files(path = "phastCons/Arabidopsis/alternative/upstream_exon/", pattern = "_phylop.wig", full.names = T)
upexon_scores <- do.call(cbind, lapply(upexon_scores, read_delim, delim = "\n"))
colnames(upexon_scores) <- alt_genes_ara
upexon_scores <- upexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
upexon_scores$med <- rowMedians(as.matrix(upexon_scores[, c(1:21)]))
median <- upexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
mean <- upexon_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
upexon_scores <- cbind(upexon_scores, median, mean)
upexon_scores$position <- c(-40:-1)

upexon_scores_c = list.files(path = "phastCons/Arabidopsis/constitutive/upstream_exon/", pattern = "_phylop.wig", full.names = T)
upexon_scores_c <- do.call(cbind, lapply(upexon_scores_c, read_delim, delim = "\n"))
colnames(upexon_scores_c) <- cons_genes_ara
upexon_scores_c <- upexon_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
upexon_scores_c$med <- rowMedians(as.matrix(upexon_scores_c[, c(1:22)]))
median_c <- upexon_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
mean_c <- upexon_scores_c$ave
mean_c <- rollapply(mean_c, width = 3, by = 1, FUN = mean, partial = T)
upexon_scores_c <- cbind(upexon_scores_c, median_c, mean_c)

upexon_scores_plot <- as.data.frame(cbind(upexon_scores$position, upexon_scores$median, upexon_scores_c$median_c, upexon_scores$mean, upexon_scores_c$mean_c))
colnames(upexon_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
upexon_scores_plot <-  upexon_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(upexon_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  lims(y = c(0.035, 0.05))+
  labs(title = "Upstream exons - Arabidopsis only",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 5' SS")+
  theme(legend.title = element_blank())

```

downstream exons
```{r}
downexon_scores = list.files(path = "phastCons/Arabidopsis/alternative/downstream_exon/", pattern = "_phylop.wig", full.names = T)
downexon_scores <- do.call(cbind, lapply(downexon_scores, read_delim, delim = "\n"))
colnames(downexon_scores) <- alt_genes_ara
downexon_scores <- downexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
downexon_scores$med <- rowMedians(as.matrix(downexon_scores[, c(1:21)]))
median <- downexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
mean <- downexon_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
downexon_scores <- cbind(downexon_scores, median, mean)
downexon_scores$position <- c(1:40)

downexon_scores_c = list.files(path = "phastCons/Arabidopsis/constitutive/downstream_exon/", pattern = "_phylop.wig", full.names = T)
downexon_scores_c <- do.call(cbind, lapply(downexon_scores_c, read_delim, delim = "\n"))
colnames(downexon_scores_c) <- cons_genes_ara
downexon_scores_c <- downexon_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
downexon_scores_c$med <- rowMedians(as.matrix(downexon_scores_c[, c(1:22)]))
median_c <- downexon_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
mean_c <- downexon_scores_c$ave
mean_c <- rollapply(mean_c, width = 3, by = 1, FUN = mean, partial = T)
downexon_scores_c <- cbind(downexon_scores_c, median_c, mean_c)

downexon_scores_plot <- as.data.frame(cbind(downexon_scores$position, downexon_scores$median, downexon_scores_c$median_c, downexon_scores$mean, downexon_scores_c$mean_c))
colnames(downexon_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
downexon_scores_plot <-  downexon_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(downexon_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  lims(y = c(0.035, 0.05))+
  labs(title = "Downstream exons - Arabidopsis only",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 3' SS")+
  theme(legend.title = element_blank())
```

intron5
```{r}
intron5_scores = list.files(path = "phastCons/Arabidopsis/alternative/intron5/", pattern = "_phylop.wig", full.names = T)
intron5_scores <- do.call(cbind, lapply(intron5_scores, read_delim, delim = "\n"))
colnames(intron5_scores) <- alt_genes_ara
intron5_scores <- intron5_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron5_scores$med <- rowMedians(as.matrix(intron5_scores[, c(1:21)]))
median <- intron5_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
mean <- intron5_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
intron5_scores <- cbind(intron5_scores, median, mean)
intron5_scores$position <- c(1:40)

intron5_scores_c = list.files(path = "phastCons/Arabidopsis/constitutive/intron5/", pattern = "_phylop.wig", full.names = T)
intron5_scores_c <- do.call(cbind, lapply(intron5_scores_c, read_delim, delim = "\n"))
colnames(intron5_scores_c) <- cons_genes_ara
intron5_scores_c <- intron5_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron5_scores_c$med <- rowMedians(as.matrix(intron5_scores_c[, c(1:22)]))
median_c <- intron5_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
mean_c <- intron5_scores_c$ave
mean_c <- rollapply(mean_c, width = 3, by = 1, FUN = mean, partial = T)
intron5_scores_c <- cbind(intron5_scores_c, median_c, mean_c)

intron5_scores_plot <- as.data.frame(cbind(intron5_scores$position, intron5_scores$median, intron5_scores_c$median_c, intron5_scores$mean, intron5_scores_c$mean_c))
colnames(intron5_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
intron5_scores_plot <-  intron5_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(intron5_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  #lims(y = c(0, 0.1))+
  labs(title = "Intron 5' end - Arabidopsis only",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 5' SS")+
  theme(legend.title = element_blank())
```

intron 3
```{r}
intron3_scores = list.files(path = "phastCons/Arabidopsis/alternative/intron3/", pattern = "_phylop.wig", full.names = T)
intron3_scores <- do.call(cbind, lapply(intron3_scores, read_delim, delim = "\n"))
colnames(intron3_scores) <- alt_genes_ara
intron3_scores <- intron3_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron3_scores$med <- rowMedians(as.matrix(intron3_scores[, c(1:21)]))
median <- intron3_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
mean <- intron3_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
intron3_scores <- cbind(intron3_scores, median, mean)
intron3_scores$position <- c(-40:-1)

intron3_scores_c = list.files(path = "phastCons/Arabidopsis/constitutive/intron3/", pattern = "_phylop.wig", full.names = T)
intron3_scores_c <- do.call(cbind, lapply(intron3_scores_c, read_delim, delim = "\n"))
colnames(intron3_scores_c) <- cons_genes_ara
intron3_scores_c <- intron3_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130))) 
intron3_scores_c$med <- rowMedians(as.matrix(intron3_scores_c[, c(1:22)]))
median_c <- intron3_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
mean_c <- intron3_scores_c$ave
mean_c <- rollapply(mean_c, width = 3, by = 1, FUN = mean, partial = T)
intron3_scores_c <- cbind(intron3_scores_c, median_c, mean_c)

intron3_scores_plot <- as.data.frame(cbind(intron3_scores$position, intron3_scores$median, intron3_scores_c$median_c, intron3_scores$mean, intron3_scores_c$mean_c))
colnames(intron3_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
intron3_scores_plot <-  intron3_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(intron3_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  #lims(y = c(0, 0.1))+
  labs(title = "Intron 3' end - Arabidopsis only",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 3' SS")+
  theme(legend.title = element_blank())
```

standalone bit of code to make phylogenetic tree and structural conservation barplot
```{r}
library(ggtree)
library(tidyverse)
```

```{r}
tree <- read.tree("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/phyloFit_tree.nwk")
tree$tip.label <- c("A. thaliana", "A. halleri", "A. lyrata", "C. rubella", "B. rapa", "S. tuberosum", "S. lycopersicum", "G. max")
tree_data <- data.frame(alt_ortho = c(23, 21, 21, 22, 22, 18, 19, 16),
                        alt_struct = c(23, 19, 21, 21, 19, 13, 15, 11),
                        const_ortho = c(23, 21, 22, 22, 22, 19, 20, 17),
                        const_struct = c(23, 18, 21, 21, 20, 15, 16, 15),
                        tip.label = c("A. thaliana", "A. halleri", "A. lyrata", "C. rubella", "B. rapa", "S. tuberosum", "S. lycopersicum", "G. max"))
tree_data <- tree_data %>% mutate(alt = alt_struct/alt_ortho, const = const_struct/const_ortho) %>% 
  pivot_longer(c(alt, const), names_to = "type", values_to = "conservation")
tree_data$tip.label <- factor(tree_data$tip.label, levels = c("A. lyrata", "A. halleri", "A. thaliana", "C. rubella", "B. rapa", "S. lycopersicum", "S. tuberosum", "G. max"))
#all_data <- full_join(tree, tree_data, by = "tip.label")
ggtree(tree)+
  geom_tiplab(hjust = 1, size = 4, vjust = -.5)+
  geom_facet(panel = "helooo", data = tree_data, geom = geom_col, aes(x = alt_ortho, colour = tip.label), orientation = "y")
#try aplot for alignment
g <- ggtree(tree, ladderize = F)+
  geom_tiplab(hjust = 1, size = 4, vjust = -.5)
b <- ggplot(tree_data, aes(x = conservation, fill = type, y = fct_rev(tip.label)))+
  geom_bar(aes(fill = type), position = position_dodge(), stat = "identity")
  #geom_text(aes(label = tip.label))
  #coord_flip()+
 # theme_tree2()
b
library(aplot)
g %>% insert_right(b)
g
```

from the 1001 genomes data, we did phyloP on 20 accessions
upstream exons
```{r}
upexon_scores = list.files(path = "phastCons/1001genomes/alternative/upstream_exon/", pattern = "_phylop.wig", full.names = T)
upexon_scores <- do.call(cbind, lapply(upexon_scores, read_delim, delim = "\n"))
colnames(upexon_scores) <- alt_genes
upexon_scores <- upexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130), na.rm = T)) 
upexon_scores$med <- rowMedians(as.matrix(upexon_scores[, c(1:23)]), na.rm = T)
median <- upexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
mean <- upexon_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
upexon_scores <- cbind(upexon_scores, median, mean)
upexon_scores$position <- c(-40:-1)

upexon_scores_c = list.files(path = "phastCons/1001genomes/constitutive/upstream_exon/", pattern = "_phylop.wig", full.names = T)
upexon_scores_c <- do.call(cbind, lapply(upexon_scores_c, read_delim, delim = "\n"))
colnames(upexon_scores_c) <- cons_genes
upexon_scores_c$AT2G45910 <- NULL #really weird numbers here throwing off mean
upexon_scores_c <- as.data.frame(sapply(upexon_scores_c, as.numeric))
#weird thing up with one of the rows in this wig file
upexon_scores_c <- upexon_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130), na.rm = T)) 
upexon_scores_c$med <- rowMedians(as.matrix(upexon_scores_c[, c(1:22)]), na.rm = T)
median_c <- upexon_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
mean_c <- upexon_scores_c$ave
mean_c <- rollapply(mean_c, width = 3, by = 1, FUN = mean, partial = T)
upexon_scores_c <- cbind(upexon_scores_c, median_c, mean_c)


upexon_scores_plot <- as.data.frame(cbind(upexon_scores$position, upexon_scores$median, upexon_scores_c$median_c, upexon_scores$mean, upexon_scores_c$mean_c))
colnames(upexon_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
upexon_scores_plot <-  upexon_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(upexon_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  #lims(y = c(0.035, 0.05))+
  labs(title = "Upstream exons - 20 Arabidopsis accessions",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 5' SS")+
  theme(legend.title = element_blank())

```

downstream exons
```{r}
downexon_scores = list.files(path = "phastCons/1001genomes/alternative/downstream_exon/", pattern = "_phylop.wig", full.names = T)
downexon_scores <- do.call(cbind, lapply(downexon_scores, read_delim, delim = "\n"))
colnames(downexon_scores) <- alt_genes
downexon_scores$AT4G34830 <- NULL
downexon_scores <- downexon_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130), na.rm = T)) 
downexon_scores$med <- rowMedians(as.matrix(downexon_scores[, c(1:22)]), na.rm = T)
median <- downexon_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
mean <- downexon_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
downexon_scores <- cbind(downexon_scores, median, mean)
downexon_scores$position <- c(1:40)

downexon_scores_c = list.files(path = "phastCons/1001genomes/constitutive/downstream_exon/", pattern = "_phylop.wig", full.names = T)
downexon_scores_c <- do.call(cbind, lapply(downexon_scores_c, read_delim, delim = "\n"))
colnames(downexon_scores_c) <- cons_genes
downexon_scores_c$AT2G45910 <- NULL #weird values again
downexon_scores_c <- as.data.frame(sapply(downexon_scores_c, as.numeric))
downexon_scores_c <- downexon_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130), na.rm = T)) 
downexon_scores_c$med <- rowMedians(as.matrix(downexon_scores_c[, c(1:22)]), na.rm = T)
median_c <- downexon_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
mean_c <- downexon_scores_c$ave
mean_c <- rollapply(mean_c, width = 3, by = 1, FUN = mean, partial = T)
downexon_scores_c <- cbind(downexon_scores_c, median_c, mean_c)

downexon_scores_plot <- as.data.frame(cbind(downexon_scores$position, downexon_scores$median, downexon_scores_c$median_c, downexon_scores$mean, downexon_scores_c$mean_c))
colnames(downexon_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
downexon_scores_plot <-  downexon_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(downexon_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  #lims(y = c(0.035, 0.05))+
  labs(title = "Downstream exons - 20 Arabidopsis accessions",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 3' SS")+
  theme(legend.title = element_blank())
```

intron 5
```{r}
intron5_scores = list.files(path = "phastCons/1001genomes/alternative/intron5/", pattern = "_phylop.wig", full.names = T)
intron5_scores <- do.call(cbind, lapply(intron5_scores, read_delim, delim = "\n"))
colnames(intron5_scores) <- alt_genes
intron5_scores$AT3G44990 <- NULL
intron5_scores <- as.data.frame(sapply(intron5_scores, as.numeric))
intron5_scores <- intron5_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130), na.rm = T)) 
intron5_scores$med <- rowMedians(as.matrix(intron5_scores[, c(1:22)]), na.rm = T)
median <- intron5_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
mean <- intron5_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
intron5_scores <- cbind(intron5_scores, median, mean)
intron5_scores$position <- c(1:40)

intron5_scores_c = list.files(path = "phastCons/1001genomes/constitutive/intron5/", pattern = "_phylop.wig", full.names = T)
intron5_scores_c <- do.call(cbind, lapply(intron5_scores_c, read_delim, delim = "\n"))
colnames(intron5_scores_c) <- cons_genes
intron5_scores_c$AT5G64130 <- NULL
intron5_scores_c$AT2G45910 <- NULL
intron5_scores_c <- as.data.frame(sapply(intron5_scores_c, as.numeric))
intron5_scores_c <- intron5_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G49940), na.rm = T)) 
intron5_scores_c$med <- rowMedians(as.matrix(intron5_scores_c[, c(1:21)]), na.rm = T)
median_c <- intron5_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
mean_c <- intron5_scores_c$ave
mean_c <- rollapply(mean_c, width = 3, by = 1, FUN = mean, partial = T)
intron5_scores_c <- cbind(intron5_scores_c, median_c, mean_c)

intron5_scores_plot <- as.data.frame(cbind(intron5_scores$position, intron5_scores$median, intron5_scores_c$median_c, intron5_scores$mean, intron5_scores_c$mean_c))
colnames(intron5_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
intron5_scores_plot <-  intron5_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(intron5_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  #lims(y = c(0, 0.1))+
  labs(title = "Intron 5' end - 20 Arabidopsis accessions",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 5' SS")+
  theme(legend.title = element_blank())
```

intron 3
```{r}
intron3_scores = list.files(path = "phastCons/1001genomes/alternative/intron3/", pattern = "_phylop.wig", full.names = T)
intron3_scores <- do.call(cbind, lapply(intron3_scores, read_delim, delim = "\n"))
colnames(intron3_scores) <- alt_genes
intron3_scores$AT3G48570 <- NULL
intron3_scores$AT3G54660 <- NULL
intron3_scores$AT5G15960 <- NULL
intron3_scores <- as.data.frame(sapply(intron3_scores, as.numeric))
intron3_scores <- intron3_scores %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130), na.rm = T)) 
intron3_scores$med <- rowMedians(as.matrix(intron3_scores[, c(1:20)]), na.rm = T)
median <- intron3_scores$med
median <- rollapply(median, width = 3, by = 1, FUN = median, partial = T)
mean <- intron3_scores$ave
mean <- rollapply(mean, width = 3, by = 1, FUN = mean, partial = T)
intron3_scores <- cbind(intron3_scores, median, mean)
intron3_scores$position <- c(-40:-1)

intron3_scores_c = list.files(path = "phastCons/1001genomes/constitutive/intron3/", pattern = "_phylop.wig", full.names = T)
intron3_scores_c <- do.call(cbind, lapply(intron3_scores_c, read_delim, delim = "\n"))
colnames(intron3_scores_c) <- cons_genes
intron3_scores_c$AT5G15960 <- NULL
intron3_scores_c$AT3G44990 <- NULL
intron3_scores_c$AT2G45910 <- NULL
intron3_scores_c <- as.data.frame(sapply(intron3_scores_c, as.numeric))
intron3_scores_c <- intron3_scores_c %>% mutate(ave = rowMeans(select(., AT1G09060:AT5G64130), na.rm = T)) 
intron3_scores_c$med <- rowMedians(as.matrix(intron3_scores_c[, c(1:20)]), na.rm = T)
median_c <- intron3_scores_c$med
median_c <- rollapply(median_c, width = 3, by = 1, FUN = median, partial = T)
mean_c <- intron3_scores_c$ave
mean_c <- rollapply(mean_c, width = 3, by = 1, FUN = mean, partial = T)
intron3_scores_c <- cbind(intron3_scores_c, median_c, mean_c)

intron3_scores_plot <- as.data.frame(cbind(intron3_scores$position, intron3_scores$median, intron3_scores_c$median_c, intron3_scores$mean, intron3_scores_c$mean_c))
colnames(intron3_scores_plot) <- c("position", "alt_med", "cons_med", "alt_mean", "cons_mean")
intron3_scores_plot <-  intron3_scores_plot %>% pivot_longer(c(alt_med, cons_med), names_to = "Type_median", values_to = "median") %>% 
  pivot_longer(c(alt_mean, cons_mean), names_to = "Type_mean", values_to = "mean")
ggplot(intron3_scores_plot, aes(x = position, y = mean, colour = Type_mean))+
  geom_line()+
  #lims(y = c(0, 0.1))+
  labs(title = "Intron 3' end - Arabidopsis only",
       y = "Mean phyloP score (3nt rolling average)",
       x = "Position relative to 3' SS")+
  theme(legend.title = element_blank())
```

