---
title: "pleiotropy"
output: html_document
---

pleiotropy analysis: has dAS evolved to regulate the expression of pleiotropic genes to minimize off-target effects? It's been shown that differential expression is constrained in pleiotropic genes so we're asking, has dAS evolved to fill that role?

```{r}
library(tidyverse)
library(moments)
library(readxl)
library(ggsignif)
library(biomartr)
library(GO.db)
```

don't run
```{r}
gene_exp <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/pleiotropy/Klepikova.tsv", delim = "\t")

#add Anna's GC data (K20w) in RPKM to the Klepikova atlas, and convert all NAs to 0
#i'm somehwat concerned about the low RPKMs in the GC data, i might repeat the analysis later, excluding these values
#gc <- read_csv("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/pleiotropy/AVERAGE_rpkm.csv") %>% select(AGI, K20w)
#gc$ID <- gc$AGI
#gc$AGI <- NULL
#gene_exp <- full_join(gene_exp, gc, by = "ID")

#ended up not including Anna's K20w samples, due to potential confounding effects (know that dAS genes have high expression in Anna's data, so this data was excluded to prevent an effect of gene exp on tissue specificity)
gene_exp[is.na(gene_exp)] <- 0
gene_exp1 <- gene_exp[1:17000,]
gene_exp2 <- gene_exp[17001:34240,]
```

use a kurtosis filter to remove housekeeping genes, as in Li et al 2012
don't run this, skip to next chunk
```{r}
gene_exp1 <- gene_exp1 %>% rowwise() %>% 
  mutate(kurt = kurtosis(c_across(2:139)),
         ans = as.numeric(anscombe.test(c_across(2:139), alternative = "l")$p.value)) %>% 
  ungroup()

gene_exp2 <- gene_exp2 %>% rowwise() %>% 
  mutate(kurt = kurtosis(c_across(2:139)),
         ans = as.numeric(anscombe.test(c_across(2:139), alternative = "l")$p.value)) %>% 
  ungroup()

gene_exp <- rbind(gene_exp1, gene_exp2)
#the alternative setting is the opposite of how I thought it would work, but setting alternative = "l" tests the alternative hypothesis that kurtosis is > 3... which is what we want, it just doesn't seem that way...

#since we've done multiple comparisons, now we need to correct them (FDR)
gene_exp$padj <- p.adjust(gene_exp$ans, "fdr")

#filter out genes with padj > 0.005 (extra-strignent cutoff imposed by Li et al due to very large sample size and lots of tissue types)
lepto <- filter(gene_exp, padj <= 0.005)

nrow(lepto) / nrow(gene_exp)
#73% of genes meet the leptokurtosis filter (not housekeeping genes). Li et al found that 62% of Arabidopsis genes were leptokurtic using microarray data, but our value is in the right ballpark and similar to values obtained for mammals
#this goes up to 74% when K20w is excluded, so not a big change

slightly_lepto <- filter(gene_exp, padj <= 0.05)
nrow(slightly_lepto) / nrow(gene_exp)
#if we use a standard 0.05 cutoff, we keep 80% of genes, so we aren't losing too many more with the strict cutoff

#write.table(lepto, "C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/pleiotropy/leptokurtic_genes_Klepikova_only.tsv", col.names = T, row.names=F, quote = F, sep = "\t")
```

```{r}
lepto <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/pleiotropy/leptokurtic_genes_Klepikova_only.tsv", delim = "\t")

gene_exp <- read_delim("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/pleiotropy/Klepikova.tsv", delim = "\t")
```

calculate tissue-specificity for the leptokurtic genes. This will distinguish between tissue-specific genes, and those that are expressed in multiple tissue types, and is the metric we'll use for comparing dAS, DE and unregulated genes, as in Rogers et al, 2021, using the tissue-specificity index developed by Yanai et al 2005. 
```{r}
tau <- function(gene){
  exp <- filter(lepto, ID == gene)
  raw_exp <- unlist(exp[2:139])
  max <- max(raw_exp)
  norm_exp <- raw_exp/max
  component <- (1 - norm_exp)/137
  tau <- sum(component, na.rm = T)
}
exp_profile <- sapply(lepto$ID, tau)
exp_profile <- as.data.frame(exp_profile)
exp_profile$ID <- row.names(exp_profile)
#kurtosis filter seems to have worked as minimum tau is 0.37, so housekeeping genes seem to have been well removed

#write.table(exp_profile, "C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/pleiotropy/exp_profile.tsv", sep = "\t", quote = F, row.names = F)

exp_profile <- read_tsv("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/conservation/pleiotropy/exp_profile.tsv")
```

now compare tau values (gene expression breadth) of non-housekeeping dAS, DE, and unregulated genes
```{r}
#dASGs
all_dasg <- read_tsv("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/clustering/vm_all_dAS_genes.tsv") %>% 
  filter(!comp %in% c("K20dL20d", "K40dL40d", "K60dL60d")) %>%
  select(GeneID)
all_dasg <- as.data.frame(unique(all_dasg$GeneID))
colnames(all_dasg) <- "ID"

all_degs <- read_tsv("C:/Users/hasna/Documents/MSc_project/majiq/dAS_analysis/clustering/dAS_DEG/all_degs.tsv")

all_unreg <- filter(gene_exp, !ID %in% all_dasg$ID & !ID %in% all_degs$ID) %>% 
  dplyr::select(ID)

```

plot tissue specificity of dAS, DE and unregulated genes
```{r}
all_dasg <- left_join(all_dasg, exp_profile) %>% 
  mutate(group = "dAS")
all_degs <- left_join(all_degs, exp_profile) %>% 
  mutate(group = "DE")
all_unreg <- left_join(all_unreg, exp_profile) %>% 
  mutate(group = "unreg")
all_taus <- rbind(all_dasg, all_degs, all_unreg) %>% 
  filter(exp_profile > 0) %>% 
  #filter out housekeeping genes, where exp_profile is na
  mutate(group = factor(group, levels = c("Unregulated",  "DE", "dAS"))) %>%
  group_by(group) %>%
  mutate(outlier = exp_profile < quantile(exp_profile, 0.25) - IQR(exp_profile) * 1.5) %>%
  ungroup

all_taus %>% group_by(group) %>% 
  summarize (n=n())
#244 dAS, 6380 DE, 18977 unregulated genes pass the leptokurtosis filter
#this represents 60% of dAS genes, 71% of DE genes and 75% of unregulated genes

model <- aov(all_taus$exp_profile ~ all_taus$group)
summary(model)
tukey <- TukeyHSD(model)
```

```{r}
ggplot(all_taus, aes(x = group, y = exp_profile, fill = group))+
  geom_boxplot(alpha = 0.6, outlier.shape = NA)+
  geom_point(data = function(x) dplyr::filter_(x, ~ outlier), position = position_jitter(width = 0.3, height = 0), alpha = 0.5, size = 0.5)+
  labs(y = "Tissue specificity index",
       x = "")+
  scale_fill_manual(values = c("orchid", "turquoise", "limegreen"))+
  theme_minimal()+
  theme(legend.position = "none")+
  scale_x_discrete(labels = c("Unregulated" = "Unregulated genes\nn = 18 977", "dAS" ="dAS genes\nn = 244", "DE" = "DE genes\n n =6380"))+
  geom_signif(y_position = c(1.05, 1.1, 1.15), xmin = c(0.6, 1.6, 0.6), xmax = c(2.4, 3.4, 3.4), annotations = c("***", "***", "***"), colour = "grey10", tip_length = 0.01, vjust = .5)
```

so those differences are highly signficant, but is there a confounding effect of TPM?
```{r}
rpkm <- lepto %>% pivot_longer(cols = 2:140, names_to = "tissue", values_to = "rpkm") %>% 
  group_by(ID) %>% 
  summarize(min = min(rpkm), max = max(rpkm), med = median(rpkm)) 
all_taus <- left_join(all_taus, rpkm)

taus_long <- all_taus %>% pivot_longer(min:med, names_to ="stat", values_to = "value")
twowaymax <- aov(all_taus$exp_profile ~ all_taus$group * all_taus$max)
summary(twowaymax)
twowaymin <- aov(all_taus$exp_profile ~ all_taus$group * all_taus$min)
summary(twowaymin)
twowaymed <- aov(all_taus$exp_profile ~ all_taus$group * all_taus$med)
summary(twowaymed)
```

the interaction between group and gene expression (min and med) is statistically significant, but is it really relevant?
```{r}
png("test.png", units="in", width=6, height=4, res=300)
ggplot(taus_long, aes(x = value, y = exp_profile, colour = group))+
  geom_point()+
  geom_smooth(method = "lm", se = F, colour = "black")+
  lims(y = c(0.25, 1))+
  facet_wrap(~group*stat, scales = "free")+
  theme(legend.position =  "none")+
  labs(y = "Tissue specificity index",
       x = "Expression level (RPKM)")+
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10))
dev.off()
```

the only one of these that look really bad is min: seems that unregulated genes have a negative correlation between minimum gene expression and tissue specificity (if min expression is low, tissue specificity is high, which makes intuitive sense, it's just much steeper in unregulated genes). So i guess the question is, do dAS genes have different minimum expression levels than DE and unregulated?
```{r}
ggplot(all_taus, aes(x = group, y = min))+
  geom_boxplot()+
  #lims(y = c(0,1))+
  labs(x = "",
       y = 'Minumum expression (RPKM)')
expcheck <- aov(min ~ group, data = all_taus)
summary(expcheck)
TukeyHSD(expcheck)
#dAS > DE > unregulated
#this would imply that unregulated genes have lower minimum expression that's driving their higher tissue specificity... should we be concerned about this? or is it that dAS genes have lower specificity, and so therefore they have higher min expression (ie which is cause vs effect). Do test this, does med expression in Anna's data (shown to be elevated in dAS genes) correlate with min Klepikova expression?

TPMs <- read_excel("C:/Users/hasna/Documents/MSc_project/20210920_rawseqdata/AVERAGE_TPM.xlsx", sheet = "AVERAGE_TPM") 
TPMs <- TPMs %>% pivot_longer(cols = K60d:L20w, names_to = "treatment", values_to = "TPM")%>% 
  group_by(AGI) %>% 
  summarize(min_Anna = min(TPM), max_Anna = max(TPM), med_Anna = median(TPM)) %>% 
  mutate(ID = AGI)

expression <- gene_exp %>% pivot_longer(cols = 2:140, names_to = "tissue", values_to = "rpkm") %>% 
  group_by(ID) %>% 
  summarize(min_Klep = min(rpkm), max_Klep = max(rpkm), med_Klep = median(rpkm)) 
expression <- full_join(expression, TPMs)
exp_model <- lm(min_Klep ~ med_Anna, data = expression)
summary(exp_model) #R^2 = 0.76
ggplot(expression, aes(x = med_Anna, y = min_Klep))+
  geom_point()+
  geom_smooth(method = "lm")

#what if we remove Anna's data from Klepikova?
expression2 <- gene_exp %>% pivot_longer(cols = 2:139, names_to = "tissue", values_to = "rpkm") %>% 
  group_by(ID) %>% 
  summarize(min_Klep2 = min(rpkm), max_Klep2 = max(rpkm), med_Klep2 = median(rpkm)) 
expression2 <- full_join(expression2, TPMs)
exp_model2 <- lm(min_Klep2 ~ med_Anna, data = expression2)
summary(exp_model2) #R^2 = 0.01
ggplot(expression2, aes(x = med_Anna, y = min_Klep2))+
  geom_point()+
  geom_smooth(method = "lm")+
  labs(x = "RPKM (Anna median)",
       y = "RPKM (Klepikova minimum")+
  lims(y=c(0, 250))

exp_model3 <- lm(med_Klep2 ~ med_Anna, data = expression2)
summary(exp_model3) #R^2 = 0.25

ggplot(expression2, aes(x = med_Anna, y = med_Klep2))+
  geom_point()+
  geom_smooth(method = "lm")+
  labs(x = "RPKM (Anna median)",
       y = "RPKM (Klepikova median)")+
  lims(y = c(0, 2500))
#MUCH WEAKER CORR!! ok now we're on to something! So in terms of the cause and effect question, expression in Anna's dataset is very weakly correlated with minimum (and median) Klepikova expression, so the reason dAS genes have higher minimum (and median) expression is because they have low specificity (and not the other way around)
```

all in all, the tissue specificity stuff is convincing. there are other proxies of pleiotropy, including go biological processes and network connectivity. let's try the go stuff:
```{r}
#no longer have a kurtosis filter, so use all genes
#get go terms from ensembl plants via biomart
go_dasg <- getGO(organism = "Arabidopsis thaliana", 
                genes    = all_dasg$ID,
                filters  = "ensembl_gene_id")
go_dasg$ont <- sapply(go_dasg$goslim_goa_accession, Ontology)
go_dasg <- go_dasg %>% group_by(ensembl_gene_id, ont) %>% 
  summarise(count = n()) %>% 
  mutate(group = c("dAS"))

go_deg <- getGO(organism = "Arabidopsis thaliana", 
                genes    = all_degs$ID,
                filters  = "ensembl_gene_id")
go_deg$ont <- sapply(go_deg$goslim_goa_accession, Ontology)
go_deg <- go_deg %>% group_by(ensembl_gene_id, ont) %>% 
  summarise(count = n()) %>% 
  mutate(group = c("DE"))

go_unreg <- getGO(organism = "Arabidopsis thaliana", 
                genes    = all_unreg$ID,
                filters  = "ensembl_gene_id")
go_unreg$ont <- sapply(go_unreg$goslim_goa_accession, Ontology)
go_unreg <- go_unreg %>% group_by(ensembl_gene_id, ont) %>% 
  summarise(count = n()) %>% 
  mutate(group = c("unreg"))

all_go <- rbind(go_dasg, go_deg, go_unreg)
#what to do about genes that don't have a go term in each ontology? Need to fill these in with zeroes somehow... 

all_go <- all_go %>% pivot_wider(names_from = ont, values_from = count) 
all_go[is.na(all_go)] <- 0

#what about genes with no go terms? add 0 for each ontology for these genes
#why doesn't biomart have annotations for these? some seem obscure, but others are annotated in tair...
no_go_dAS <- filter(all_dasg, !ID %in% go_dasg$ensembl_gene_id)
no_go_de <- filter(all_degs, !ID %in% go_deg$ensembl_gene_id)
no_go_unreg <- filter(all_unreg, !ID %in% go_unreg$ensembl_gene_id)
no_go <- rbind(no_go_dAS, no_go_de, no_go_unreg)
no_go <- no_go[,-2]
no_go$BP <- c(0)
no_go$CC <- c(0)
no_go$MF <- c(0)
colnames(no_go) <- c("ensembl_gene_id", "group", "BP", "CC", "MF")

all_go <- rbind(all_go, no_go)

all_go %>% group_by(group) %>% 
  summarize (n=n())

all_go <- pivot_longer(all_go, BP:MF, names_to = "ont", values_to = "count")

bp <- filter(all_go, ont == "BP") %>% 
  mutate(group = factor(group, levels = c("unreg",  "DE", "dAS"))) %>%
  group_by(group) %>%
  mutate(outlier = count > quantile(count, 0.75) + IQR(count) * 1.5 | count < quantile(count, 0.25) - IQR(count) * 1.5 ) %>%
  ungroup

model_bp <- aov(bp$count ~ bp$group)
summary(model_bp)
tukey_bp <- TukeyHSD(model_bp)
tukey_bp

cc <- filter(all_go, ont == "CC") %>% 
  mutate(group = factor(group, levels = c("unreg",  "DE", "dAS"))) %>%
  group_by(group) %>%
  mutate(outlier = count > quantile(count, 0.75) + IQR(count) * 1.5 | count < quantile(count, 0.25) - IQR(count) * 1.5 ) %>%
  ungroup

model_cc <- aov(cc$count ~ cc$group)
summary(model_cc)
tukey_cc <- TukeyHSD(model_cc)
tukey_cc

mf <- filter(all_go, ont == "MF") %>% 
  mutate(group = factor(group, levels = c("unreg",  "DE", "dAS"))) %>%
  group_by(group) %>%
  mutate(outlier = count > quantile(count, 0.75) + IQR(count) * 1.5 | count < quantile(count, 0.25) - IQR(count) * 1.5 ) %>%
  ungroup

model_mf <- aov(mf$count ~ mf$group)
summary(model_mf)
tukey_mf <- TukeyHSD(model_mf)
tukey_mf
```

```{r}
ggplot(bp, aes(x = group, y = count, fill = group))+
  geom_boxplot(alpha = 0.6, outlier.shape = NA)+
  geom_point(data = function(x) dplyr::filter_(x, ~ outlier), position = position_jitter(width = 0.3, height = 0), alpha = 0.5, size = 0.5)+
  labs(y = "GO biological processes per gene",
       x = "")+
  scale_fill_manual(values = c("orchid", "turquoise", "limegreen"))+
  theme_minimal()+
  theme(legend.position = "none")+
  scale_x_discrete(labels = c("unreg" = "Unregulated genes\nn = 25 153", "dAS" ="dAS genes\nn = 406", "DE" = "DE genes\n n =8989"))+
  geom_signif(y_position = c(29, 30.5, 32), xmin = c(0.6, 1.6, 0.6), xmax = c(2.4, 3.4, 3.4), annotations = c("***", "**", "***"), colour = "grey10", tip_length = 0.01, vjust = .5)
```

```{r}
ggplot(cc, aes(x = group, y = count, fill = group))+
  geom_boxplot(alpha = 0.6, outlier.shape = NA)+
  geom_point(data = function(x) dplyr::filter_(x, ~ outlier), position = position_jitter(width = 0.3, height = 0), alpha = 0.5, size = 0.5)+
  labs(y = "GO cellular components per gene",
       x = "")+
  scale_fill_manual(values = c("orchid", "turquoise", "limegreen"))+
  theme_minimal()+
  theme(legend.position = "none")+
  scale_x_discrete(labels = c("unreg" = "Unregulated genes\nn = 25 153", "dAS" ="dAS genes\nn = 406", "DE" = "DE genes\n n =8989"))+
  geom_signif(y_position = c(17, 18.25, 19.5), xmin = c(0.6, 1.6, 0.6), xmax = c(2.4, 3.4, 3.4), annotations = c("***", "***", "***"), colour = "grey10", tip_length = 0.01, vjust = .5)
```

```{r}
ggplot(mf, aes(x = group, y = count, fill = group))+
  geom_boxplot(alpha = 0.6, outlier.shape = NA)+
  geom_point(data = function(x) dplyr::filter_(x, ~ outlier), position = position_jitter(width = 0.3, height = 0), alpha = 0.5, size = 0.5)+
  labs(y = "GO molecular functions per gene",
       x = "")+
  scale_fill_manual(values = c("orchid", "turquoise", "limegreen"))+
  theme_minimal()+
  theme(legend.position = "none")+
  scale_x_discrete(labels = c("unreg" = "Unregulated genes\nn = 25 153", "dAS" ="dAS genes\nn = 406", "DE" = "DE genes\n n =8989"))+
  geom_signif(y_position = c(12.5, 13.25, 14), xmin = c(0.6, 1.6, 0.6), xmax = c(2.4, 3.4, 3.4), annotations = c("***", "NS", "***"), colour = "grey10", tip_length = 0.01, vjust = c(0.1))
```